--- a/net/minecraft/client/resources/language/ClientLanguage.java
+++ b/net/minecraft/client/resources/language/ClientLanguage.java
@@ -16,6 +16,7 @@
 import net.optifine.Lang;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.vivecraft.utils.LangHelper;
 
 public class ClientLanguage extends Language
 {
@@ -46,13 +47,16 @@
                     ResourceLocation resourcelocation = new ResourceLocation(s1, s);
                     appendFrom(p_118917_.getResources(resourcelocation), map);
                     Lang.loadResources(p_118917_, languageinfo.getCode(), map);
+                    //Vivecraft
+                    LangHelper.loadLocaleData(languageinfo.getCode(), map);
+                    //
                 }
                 catch (FileNotFoundException filenotfoundexception)
                 {
                 }
-                catch (Exception exception1)
+                catch (Exception exception)
                 {
-                    LOGGER.warn("Skipped language file: {}:{} ({})", s1, s, exception1.toString());
+                    LOGGER.warn("Skipped language file: {}:{} ({})", s1, s, exception.toString());
                 }
             }
         }
@@ -94,9 +98,9 @@
                     inputstream.close();
                 }
             }
-            catch (IOException ioexception1)
+            catch (IOException ioexception)
             {
-                LOGGER.warn("Failed to load translations from {}", resource, ioexception1);
+                LOGGER.warn("Failed to load translations from {}", resource, ioexception);
             }
         }
     }
