--- a/net/minecraft/util/profiling/ActiveProfiler.java
+++ b/net/minecraft/util/profiling/ActiveProfiler.java
@@ -135,6 +135,7 @@
 
             this.path = this.path + pName;
             this.paths.add(this.path);
+            //LOGGER.error("+" + this.path);
             this.startTimes.add(Util.getNanos());
             this.currentEntry = null;
         }
@@ -164,6 +165,7 @@
         {
             long i = Util.getNanos();
             long j = this.startTimes.removeLong(this.startTimes.size() - 1);
+            //LOGGER.error("-" + this.paths.get(this.paths.size() - 1));
             this.paths.remove(this.paths.size() - 1);
             long k = i - j;
             ActiveProfiler.PathEntry activeprofiler$pathentry = this.getCurrentEntry();
