--- a/net/minecraft/world/item/CrossbowItem.java
+++ b/net/minecraft/world/item/CrossbowItem.java
@@ -31,6 +31,9 @@
 import net.minecraft.world.item.enchantment.Enchantments;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.Vec3;
+import org.vivecraft.api.NetworkHelper;
+import org.vivecraft.api.ServerVivePlayer;
+import org.vivecraft.utils.math.Vector3;
 
 public class CrossbowItem extends ProjectileWeaponItem implements Vanishable
 {
@@ -270,6 +273,15 @@
                 Vec3 vec31 = pShooter.getUpVector(1.0F);
                 Quaternion quaternion = new Quaternion(new Vector3f(vec31), pProjectileAngle, true);
                 Vec3 vec3 = pShooter.getViewVector(1.0F);
+                //Vivecraft
+                ServerVivePlayer serverviveplayer = NetworkHelper.vivePlayers.get(pShooter.getUUID());
+
+                if (serverviveplayer != null && serverviveplayer.isVR())
+                {
+                    vec3 = serverviveplayer.getControllerDir(serverviveplayer.activeHand);
+                    serverviveplayer.getControllerVectorCustom(serverviveplayer.activeHand, new Vector3(0.0F, 1.0F, 0.0F));
+                }
+                //
                 Vector3f vector3f = new Vector3f(vec3);
                 vector3f.transform(quaternion);
                 projectile.shoot((double)vector3f.x(), (double)vector3f.y(), (double)vector3f.z(), pVelocity, pInaccuracy);
@@ -428,7 +440,7 @@
         }
     }
 
-    private static float getPowerForTime(int pUseTime, ItemStack pCrossbowStack)
+    public static float getPowerForTime(int pUseTime, ItemStack pCrossbowStack)
     {
         float f = (float)pUseTime / (float)getChargeDuration(pCrossbowStack);
 
