--- a/com/mojang/blaze3d/platform/GlStateManager.java
+++ b/com/mojang/blaze3d/platform/GlStateManager.java
@@ -42,7 +42,7 @@
 {
     private static final boolean ON_LINUX = Util.getPlatform() == Util.OS.LINUX;
     public static final int TEXTURE_COUNT = 12;
-    private static final GlStateManager.BlendState BLEND = new GlStateManager.BlendState();
+    public static final GlStateManager.BlendState BLEND = new GlStateManager.BlendState();
     private static final GlStateManager.DepthState DEPTH = new GlStateManager.DepthState();
     private static final GlStateManager.CullState CULL = new GlStateManager.CullState();
     private static final GlStateManager.PolygonOffsetState POLY_OFFSET = new GlStateManager.PolygonOffsetState();
@@ -250,26 +250,30 @@
     {
         RenderSystem.assertOnRenderThread();
 
+        // VIVECRAFT: correct bad blend function that trashes alpha channel
+        if (pSrcFactor == GlStateManager.SourceFactor.SRC_ALPHA.value && pDstFactor == GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA.value && pSrcFactorAlpha == GlStateManager.SourceFactor.ONE.value && pDstFactorAlpha == GlStateManager.DestFactor.ZERO.value)
+        {
+            pDstFactorAlpha = GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA.value;
+        }
+        //
+
         if (blendLock.isLocked())
         {
             blendLockState.setFactors(pSrcFactor, pDstFactor, pSrcFactorAlpha, pDstFactorAlpha);
         }
-        else
+        else if (pSrcFactor != BLEND.srcRgb || pDstFactor != BLEND.dstRgb || pSrcFactorAlpha != BLEND.srcAlpha || pDstFactorAlpha != BLEND.dstAlpha)
         {
-            if (pSrcFactor != BLEND.srcRgb || pDstFactor != BLEND.dstRgb || pSrcFactorAlpha != BLEND.srcAlpha || pDstFactorAlpha != BLEND.dstAlpha)
-            {
-                BLEND.srcRgb = pSrcFactor;
-                BLEND.dstRgb = pDstFactor;
-                BLEND.srcAlpha = pSrcFactorAlpha;
-                BLEND.dstAlpha = pDstFactorAlpha;
-
-                if (Config.isShaders())
-                {
-                    Shaders.uniform_blendFunc.setValue(pSrcFactor, pDstFactor, pSrcFactorAlpha, pDstFactorAlpha);
-                }
+            BLEND.srcRgb = pSrcFactor;
+            BLEND.dstRgb = pDstFactor;
+            BLEND.srcAlpha = pSrcFactorAlpha;
+            BLEND.dstAlpha = pDstFactorAlpha;
 
-                glBlendFuncSeparate(pSrcFactor, pDstFactor, pSrcFactorAlpha, pDstFactorAlpha);
+            if (Config.isShaders())
+            {
+                Shaders.uniform_blendFunc.setValue(pSrcFactor, pDstFactor, pSrcFactorAlpha, pDstFactorAlpha);
             }
+
+            glBlendFuncSeparate(pSrcFactor, pDstFactor, pSrcFactorAlpha, pDstFactorAlpha);
         }
     }
 
@@ -904,11 +908,11 @@
 
             GL11.glDeleteTextures(pTexture);
 
-            for (GlStateManager.TextureState glstatemanager$texturestate : TEXTURES)
+            for (GlStateManager.TextureState GlStateManager$texturestate : TEXTURES)
             {
-                if (glstatemanager$texturestate.binding == pTexture)
+                if (GlStateManager$texturestate.binding == pTexture)
                 {
-                    glstatemanager$texturestate.binding = 0;
+                    GlStateManager$texturestate.binding = 0;
                 }
             }
         }
@@ -1299,45 +1303,45 @@
         }
     }
 
-    public static void getBlendState(GlBlendState gbs)
+    public static void getBlendState(GlBlendState p_getBlendState_0_)
     {
         if (blendLock.isLocked())
         {
-            gbs.setState(blendLockState);
+            p_getBlendState_0_.setState(blendLockState);
         }
         else
         {
-            gbs.setState(BLEND.mode.enabled, BLEND.srcRgb, BLEND.dstRgb, BLEND.srcAlpha, BLEND.dstAlpha);
+            p_getBlendState_0_.setState(BLEND.mode.enabled, BLEND.srcRgb, BLEND.dstRgb, BLEND.srcAlpha, BLEND.dstAlpha);
         }
     }
 
-    public static void setBlendState(GlBlendState gbs)
+    public static void setBlendState(GlBlendState p_setBlendState_0_)
     {
         if (blendLock.isLocked())
         {
-            blendLockState.setState(gbs);
+            blendLockState.setState(p_setBlendState_0_);
         }
         else
         {
-            BLEND.mode.setEnabled(gbs.isEnabled());
+            BLEND.mode.setEnabled(p_setBlendState_0_.isEnabled());
 
-            if (!gbs.isSeparate())
+            if (!p_setBlendState_0_.isSeparate())
             {
-                _blendFunc(gbs.getSrcFactor(), gbs.getDstFactor());
+                _blendFunc(p_setBlendState_0_.getSrcFactor(), p_setBlendState_0_.getDstFactor());
             }
             else
             {
-                _blendFuncSeparate(gbs.getSrcFactor(), gbs.getDstFactor(), gbs.getSrcFactorAlpha(), gbs.getDstFactorAlpha());
+                _blendFuncSeparate(p_setBlendState_0_.getSrcFactor(), p_setBlendState_0_.getDstFactor(), p_setBlendState_0_.getSrcFactorAlpha(), p_setBlendState_0_.getDstFactorAlpha());
             }
         }
     }
 
-    public static void lockCull(GlCullState stateNew)
+    public static void lockCull(GlCullState p_lockCull_0_)
     {
         if (!cullLock.isLocked())
         {
             getCullState(cullLockState);
-            setCullState(stateNew);
+            setCullState(p_lockCull_0_);
             cullLock.lock();
         }
     }
@@ -1350,34 +1354,34 @@
         }
     }
 
-    public static void getCullState(GlCullState state)
+    public static void getCullState(GlCullState p_getCullState_0_)
     {
         if (cullLock.isLocked())
         {
-            state.setState(cullLockState);
+            p_getCullState_0_.setState(cullLockState);
         }
         else
         {
-            state.setState(CULL.enable.enabled, CULL.mode);
+            p_getCullState_0_.setState(CULL.enable.enabled, CULL.mode);
         }
     }
 
-    public static void setCullState(GlCullState state)
+    public static void setCullState(GlCullState p_setCullState_0_)
     {
         if (cullLock.isLocked())
         {
-            cullLockState.setState(state);
+            cullLockState.setState(p_setCullState_0_);
         }
         else
         {
-            CULL.enable.setEnabled(state.isEnabled());
-            CULL.mode = state.getMode();
+            CULL.enable.setEnabled(p_setCullState_0_.isEnabled());
+            CULL.mode = p_setCullState_0_.getMode();
         }
     }
 
-    public static void glMultiDrawArrays(int mode, IntBuffer bFirst, IntBuffer bCount)
+    public static void glMultiDrawArrays(int p_glMultiDrawArrays_0_, IntBuffer p_glMultiDrawArrays_1_, IntBuffer p_glMultiDrawArrays_2_)
     {
-        GL14.glMultiDrawArrays(mode, bFirst, bCount);
+        GL14.glMultiDrawArrays(p_glMultiDrawArrays_0_, p_glMultiDrawArrays_1_, p_glMultiDrawArrays_2_);
 
         if (Config.isShaders())
         {
@@ -1388,7 +1392,7 @@
                 for (int j = 1; j < i; ++j)
                 {
                     Shaders.uniform_instanceId.setValue(j);
-                    GL14.glMultiDrawArrays(mode, bFirst, bCount);
+                    GL14.glMultiDrawArrays(p_glMultiDrawArrays_0_, p_glMultiDrawArrays_1_, p_glMultiDrawArrays_2_);
                 }
 
                 Shaders.uniform_instanceId.setValue(0);
@@ -1433,9 +1437,9 @@
         GL31.glCopyBufferSubData(readTarget, writeTarget, readOffset, writeOffset, size);
     }
 
-    public static void readPixels(int x, int y, int width, int height, int format, int type, long pixels)
+    public static void readPixels(int p_readPixels_0_, int p_readPixels_1_, int p_readPixels_2_, int p_readPixels_3_, int p_readPixels_4_, int p_readPixels_5_, long p_readPixels_6_)
     {
-        GL11.glReadPixels(x, y, width, height, format, type, pixels);
+        GL11.glReadPixels(p_readPixels_0_, p_readPixels_1_, p_readPixels_2_, p_readPixels_3_, p_readPixels_4_, p_readPixels_5_, p_readPixels_6_);
     }
 
     public static int getFramebufferRead()
@@ -1462,13 +1466,13 @@
         GL14.glBlendFuncSeparate(BLEND.srcRgb, BLEND.dstRgb, BLEND.srcAlpha, BLEND.dstAlpha);
     }
 
-    public static void setBlendsIndexed(GlBlendState[] blends)
+    public static void setBlendsIndexed(GlBlendState[] p_setBlendsIndexed_0_)
     {
-        if (blends != null)
+        if (p_setBlendsIndexed_0_ != null)
         {
-            for (int i = 0; i < blends.length; ++i)
+            for (int i = 0; i < p_setBlendsIndexed_0_.length; ++i)
             {
-                GlBlendState glblendstate = blends[i];
+                GlBlendState glblendstate = p_setBlendsIndexed_0_[i];
 
                 if (glblendstate != null)
                 {
@@ -1487,19 +1491,19 @@
         }
     }
 
-    public static void bindImageTexture(int unit, int texture, int level, boolean layered, int layer, int access, int format)
+    public static void bindImageTexture(int p_bindImageTexture_0_, int p_bindImageTexture_1_, int p_bindImageTexture_2_, boolean p_bindImageTexture_3_, int p_bindImageTexture_4_, int p_bindImageTexture_5_, int p_bindImageTexture_6_)
     {
-        if (unit >= 0 && unit < IMAGE_TEXTURES.length)
+        if (p_bindImageTexture_0_ >= 0 && p_bindImageTexture_0_ < IMAGE_TEXTURES.length)
         {
-            if (IMAGE_TEXTURES[unit] == texture)
+            if (IMAGE_TEXTURES[p_bindImageTexture_0_] == p_bindImageTexture_1_)
             {
                 return;
             }
 
-            IMAGE_TEXTURES[unit] = texture;
+            IMAGE_TEXTURES[p_bindImageTexture_0_] = p_bindImageTexture_1_;
         }
 
-        GL42.glBindImageTexture(unit, texture, level, layered, layer, access, format);
+        GL42.glBindImageTexture(p_bindImageTexture_0_, p_bindImageTexture_1_, p_bindImageTexture_2_, p_bindImageTexture_3_, p_bindImageTexture_4_, p_bindImageTexture_5_, p_bindImageTexture_6_);
     }
 
     public static int getProgram()
@@ -1507,7 +1511,7 @@
         return glProgram;
     }
 
-    static class BlendState
+    public static class BlendState
     {
         public final GlStateManager.BooleanState mode = new GlStateManager.BooleanState(3042);
         public int srcRgb = 1;
@@ -1727,4 +1731,46 @@
             return INSTANCE.height;
         }
     }
+    
+    //Vivecraft TODO: fix
+    public static void mulTextureByModelView()
+    {
+//    	getMatrix(2982, MATRIX_BUFFER);
+//    	multMatrix(MATRIX_BUFFER);
+    }
+    
+    public static Matrix4f getMatrix4f(int p_getMatrix4f_0_)
+    {
+//    	GL11.glGetFloatv(p_getMatrix4f_0_, MATRIX_BUFFER);
+//    	MATRIX_BUFFER.rewind();
+//    	Matrix4f matrix4f = new Matrix4f();
+//    	matrix4f.read(MATRIX_BUFFER);
+//    	MATRIX_BUFFER.rewind();
+    	return null;
+    }
+        
+    //Vivecraft yea well un-deprecate it.
+    @Deprecated
+	public
+    static class Color
+    {
+        public float r = 1.0F;
+        public float g = 1.0F;
+        public float b = 1.0F;
+        public float a = 1.0F;
+
+        public Color()
+        {
+            this(1.0F, 1.0F, 1.0F, 1.0F);
+        }
+
+        public Color(float red, float green, float blue, float alpha)
+        {
+            this.r = red;
+            this.g = green;
+            this.b = blue;
+            this.a = alpha;
+        }
+    }
+    
 }
