--- a/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
+++ b/net/minecraft/client/renderer/block/LiquidBlockRenderer.java
@@ -36,7 +36,7 @@
     private final TextureAtlasSprite[] lavaIcons = new TextureAtlasSprite[2];
     private final TextureAtlasSprite[] waterIcons = new TextureAtlasSprite[2];
     private TextureAtlasSprite waterOverlay;
-
+    public static boolean skipStupidGoddamnChunkBoundaryClipping;
     protected void setupSprites()
     {
         this.lavaIcons[0] = Minecraft.getInstance().getModelManager().getBlockModelShaper().getBlockModel(Blocks.LAVA.defaultBlockState()).getParticleIcon();
@@ -152,9 +152,13 @@
                 float f8 = this.getWaterHeight(pLevel, pPos.south(), pFluidState.getType());
                 float f9 = this.getWaterHeight(pLevel, pPos.east().south(), pFluidState.getType());
                 float f10 = this.getWaterHeight(pLevel, pPos.east(), pFluidState.getType());
-                double d0 = (double)(pPos.getX() & 15);
-                double d1 = (double)(pPos.getY() & 15);
-                double d2 = (double)(pPos.getZ() & 15);
+                //double d0 = (double)(pPos.getX() & 15);
+                //double d1 = (double)(pPos.getY() & 15);
+                //double d2 = (double)(pPos.getZ() & 15);
+                double d0 = skipStupidGoddamnChunkBoundaryClipping ? pPos.getX() : (double)(pPos.getX() & 15);
+                double d1 = skipStupidGoddamnChunkBoundaryClipping ? pPos.getY() : (double)(pPos.getY() & 15);
+                double d2 = skipStupidGoddamnChunkBoundaryClipping ? pPos.getZ() : (double)(pPos.getZ() & 15);
+
 
                 if (Config.isRenderRegions())
                 {
