--- a/net/minecraft/client/gui/screens/inventory/BookEditScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/BookEditScreen.java
@@ -44,6 +44,7 @@
 import org.apache.commons.lang3.StringUtils;
 import org.apache.commons.lang3.mutable.MutableBoolean;
 import org.apache.commons.lang3.mutable.MutableInt;
+import org.vivecraft.gameplay.screenhandlers.KeyboardHandler;
 
 public class BookEditScreen extends Screen
 {
@@ -64,19 +65,19 @@
     private int currentPage;
     private final List<String> pages = Lists.newArrayList();
     private String title = "";
-    private final TextFieldHelper pageEdit = new TextFieldHelper(this::getCurrentPageText, this::setCurrentPageText, this::getClipboard, this::setClipboard, (p_98179_) ->
+    private final TextFieldHelper pageEdit = new TextFieldHelper(this::getCurrentPageText, this::setCurrentPageText, this::getClipboard, this::setClipboard, (p_238774_1_) ->
     {
-        return p_98179_.length() < 1024 && this.font.wordWrapHeight(p_98179_, 114) <= 128;
+        return p_238774_1_.length() < 1024 && this.font.wordWrapHeight(p_238774_1_, 114) <= 128;
     });
     private final TextFieldHelper titleEdit = new TextFieldHelper(() ->
     {
         return this.title;
-    }, (p_98175_) ->
+    }, (p_238772_1_) ->
     {
-        this.title = p_98175_;
-    }, this::getClipboard, this::setClipboard, (p_98170_) ->
+        this.title = p_238772_1_;
+    }, this::getClipboard, this::setClipboard, (p_238771_0_) ->
     {
-        return p_98170_.length() < 16;
+        return p_238771_0_.length() < 16;
     });
     private long lastClickTime;
     private int lastIndex = -1;
@@ -178,6 +179,9 @@
         {
             this.pageBack();
         }, true));
+        //Vivecraft
+        KeyboardHandler.setOverlayShowing(true);
+        //
         this.updateButtonVisibility();
     }
 
@@ -687,16 +691,16 @@
             MutableInt mutableint = new MutableInt();
             MutableBoolean mutableboolean = new MutableBoolean();
             StringSplitter stringsplitter = this.font.getSplitter();
-            stringsplitter.splitLines(s, 114, Style.EMPTY, true, (p_98132_, p_98133_, p_98134_) ->
+            stringsplitter.splitLines(s, 114, Style.EMPTY, true, (p_238762_6_, p_238762_7_, p_238762_8_) ->
             {
                 int k3 = mutableint.getAndIncrement();
-                String s2 = s.substring(p_98133_, p_98134_);
+                String s2 = s.substring(p_238762_7_, p_238762_8_);
                 mutableboolean.setValue(s2.endsWith("\n"));
                 String s3 = StringUtils.stripEnd(s2, " \n");
                 int l3 = k3 * 9;
                 BookEditScreen.Pos2i bookeditscreen$pos2i1 = this.convertLocalToScreen(new BookEditScreen.Pos2i(0, l3));
-                intlist.add(p_98133_);
-                list.add(new BookEditScreen.LineInfo(p_98132_, s3, bookeditscreen$pos2i1.x, bookeditscreen$pos2i1.y));
+                intlist.add(p_238762_7_);
+                list.add(new BookEditScreen.LineInfo(p_238762_6_, s3, bookeditscreen$pos2i1.x, bookeditscreen$pos2i1.y));
             });
             int[] aint = intlist.toIntArray();
             boolean flag = i == s.length();
