--- a/net/minecraft/client/renderer/entity/EntityRenderers.java
+++ b/net/minecraft/client/renderer/entity/EntityRenderers.java
@@ -24,12 +24,12 @@
     private static final Map<String, EntityRendererProvider<AbstractClientPlayer>> PLAYER_PROVIDERS = ImmutableMap.of("default", (p_174097_0_) ->
     {
         PlayerRenderer playerrenderer = new PlayerRenderer(p_174097_0_, false);
-        playerrenderer.addLayer(new PlayerItemsLayer(playerrenderer));
+        playerrenderer.addLayer(new PlayerItemsLayer((LivingEntityRenderer)playerrenderer));
         return playerrenderer;
     }, "slim", (p_174095_0_) ->
     {
         PlayerRenderer playerrenderer = new PlayerRenderer(p_174095_0_, true);
-        playerrenderer.addLayer(new PlayerItemsLayer(playerrenderer));
+        playerrenderer.addLayer(new PlayerItemsLayer((LivingEntityRenderer)playerrenderer));
         return playerrenderer;
     });
 
