--- a/net/minecraft/client/renderer/entity/FishingHookRenderer.java
+++ b/net/minecraft/client/renderer/entity/FishingHookRenderer.java
@@ -14,8 +14,8 @@
 import net.minecraft.world.entity.HumanoidArm;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.entity.projectile.FishingHook;
+import net.minecraft.world.item.FishingRodItem;
 import net.minecraft.world.item.ItemStack;
-import net.minecraft.world.item.Items;
 import net.minecraft.world.phys.Vec3;
 
 public class FishingHookRenderer extends EntityRenderer<FishingHook>
@@ -52,7 +52,7 @@
             int i = player.getMainArm() == HumanoidArm.RIGHT ? 1 : -1;
             ItemStack itemstack = player.getMainHandItem();
 
-            if (!itemstack.is(Items.FISHING_ROD))
+            if (!(itemstack.getItem() instanceof FishingRodItem))
             {
                 i = -i;
             }
@@ -80,6 +80,19 @@
                 d5 = Mth.lerp((double)pPartialTicks, player.yo, player.getY()) + vec3.y;
                 d6 = Mth.lerp((double)pPartialTicks, player.zo, player.getZ()) + vec3.z;
                 f3 = player.getEyeHeight();
+                int j = 1;
+
+                if (player.getMainHandItem().getItem() instanceof FishingRodItem)
+                {
+                    j = 0;
+                }
+
+                Vec3 vec31 = Minecraft.getInstance().gameRenderer.getControllerRenderPos(j);
+                Vec3 vec32 = Minecraft.getInstance().vrPlayer.vrdata_world_render.getHand(j).getDirection();
+                d4 = vec31.x + vec32.x * (double)0.47F;
+                d5 = vec31.y + vec32.y * (double)0.47F;
+                d6 = vec31.z + vec32.z * (double)0.47F;
+                f3 = 0.0F;
             }
             else
             {
