--- a/net/minecraft/world/item/Item.java
+++ b/net/minecraft/world/item/Item.java
@@ -8,6 +8,9 @@
 import java.util.Optional;
 import java.util.UUID;
 import javax.annotation.Nullable;
+
+import org.vivecraft.asm.ASMDelegator;
+
 import net.minecraft.SharedConstants;
 import net.minecraft.Util;
 import net.minecraft.core.BlockPos;
@@ -143,7 +146,7 @@
         {
             ItemStack itemstack = pPlayer.getItemInHand(pUsedHand);
 
-            if (pPlayer.canEat(this.getFoodProperties().canAlwaysEat()))
+            if (pPlayer.canEat(this.getFoodProperties().canAlwaysEat()) || itemstack.getHoverName().getString().equals("EAT ME"))
             {
                 pPlayer.startUsingItem(pUsedHand);
                 return InteractionResultHolder.consume(itemstack);
@@ -358,6 +361,15 @@
         float f = pPlayer.getXRot();
         float f1 = pPlayer.getYRot();
         Vec3 vec3 = pPlayer.getEyePosition();
+
+        // Vivecraft - Identical code to asm for vanilla and so we understand what's going on.
+        // TODO: Actually make the asm work again when we fix Forge
+        f = ASMDelegator.itemRayTracePitch(pPlayer, f);
+        f1 = ASMDelegator.itemRayTraceYaw(pPlayer, f1);
+        vec3 = ASMDelegator.itemRayTracePos(pPlayer, vec3);
+        System.out.println(Thread.currentThread().getName() + " " + vec3.x + " " + vec3.y + " " + vec3.z + " " + f + " " + f1);
+        //
+
         float f2 = Mth.cos(-f1 * ((float)Math.PI / 180F) - (float)Math.PI);
         float f3 = Mth.sin(-f1 * ((float)Math.PI / 180F) - (float)Math.PI);
         float f4 = -Mth.cos(-f * ((float)Math.PI / 180F));
