--- a/net/minecraft/client/tutorial/CraftPlanksTutorialStep.java
+++ b/net/minecraft/client/tutorial/CraftPlanksTutorialStep.java
@@ -36,7 +36,7 @@
         {
             if (this.timeWaiting == 1)
             {
-                LocalPlayer localplayer = this.tutorial.getMinecraft().player;
+                LocalPlayer localplayer = this.tutorial.getInstance().player;
 
                 if (localplayer != null)
                 {
@@ -57,7 +57,7 @@
             if (this.timeWaiting >= 1200 && this.toast == null)
             {
                 this.toast = new TutorialToast(TutorialToast.Icons.WOODEN_PLANKS, CRAFT_TITLE, CRAFT_DESCRIPTION, false);
-                this.tutorial.getMinecraft().getToasts().addToast(this.toast);
+                this.tutorial.getInstance().getToasts().addToast(this.toast);
             }
         }
     }
