--- a/net/minecraft/client/tutorial/FindTreeTutorialStepInstance.java
+++ b/net/minecraft/client/tutorial/FindTreeTutorialStepInstance.java
@@ -42,7 +42,7 @@
         {
             if (this.timeWaiting == 1)
             {
-                LocalPlayer localplayer = this.tutorial.getMinecraft().player;
+                LocalPlayer localplayer = this.tutorial.getInstance().player;
 
                 if (localplayer != null)
                 {
@@ -66,7 +66,7 @@
             if (this.timeWaiting >= 6000 && this.toast == null)
             {
                 this.toast = new TutorialToast(TutorialToast.Icons.TREE, TITLE, DESCRIPTION, false);
-                this.tutorial.getMinecraft().getToasts().addToast(this.toast);
+                this.tutorial.getInstance().getToasts().addToast(this.toast);
             }
         }
     }
