--- a/net/minecraft/client/tutorial/PunchTreeTutorialStepInstance.java
+++ b/net/minecraft/client/tutorial/PunchTreeTutorialStepInstance.java
@@ -38,7 +38,7 @@
         {
             if (this.timeWaiting == 1)
             {
-                LocalPlayer localplayer = this.tutorial.getMinecraft().player;
+                LocalPlayer localplayer = this.tutorial.getInstance().player;
 
                 if (localplayer != null)
                 {
@@ -59,7 +59,7 @@
             if ((this.timeWaiting >= 600 || this.resetCount > 3) && this.toast == null)
             {
                 this.toast = new TutorialToast(TutorialToast.Icons.TREE, TITLE, DESCRIPTION, true);
-                this.tutorial.getMinecraft().getToasts().addToast(this.toast);
+                this.tutorial.getInstance().getToasts().addToast(this.toast);
             }
         }
     }
