--- a/net/minecraft/client/gui/components/events/ContainerEventHandler.java
+++ b/net/minecraft/client/gui/components/events/ContainerEventHandler.java
@@ -6,6 +6,7 @@
 import java.util.function.BooleanSupplier;
 import java.util.function.Supplier;
 import javax.annotation.Nullable;
+import net.minecraft.client.Minecraft;
 
 public interface ContainerEventHandler extends GuiEventListener
 {
@@ -47,9 +48,9 @@
 default boolean mouseReleased(double pMouseX, double p_94723_, int pMouseY)
     {
         this.setDragging(false);
-        return this.getChildAt(pMouseX, p_94723_).filter((p_94708_) ->
+        return this.getChildAt(pMouseX, p_94723_).filter((listener) ->
         {
-            return p_94708_.mouseReleased(pMouseX, p_94723_, pMouseY);
+            return listener.mouseReleased(pMouseX, p_94723_, pMouseY);
         }).isPresent();
     }
 
@@ -64,9 +65,9 @@
 
 default boolean mouseScrolled(double pMouseX, double p_94687_, double pMouseY)
     {
-        return this.getChildAt(pMouseX, p_94687_).filter((p_94693_) ->
+        return this.getChildAt(pMouseX, p_94687_).filter((listener) ->
         {
-            return p_94693_.mouseScrolled(pMouseX, p_94687_, pMouseY);
+            return listener.mouseScrolled(pMouseX, p_94687_, pMouseY);
         }).isPresent();
     }
 
@@ -92,8 +93,13 @@
 
 default void setInitialFocus(@Nullable GuiEventListener pEventListener)
     {
-        this.setFocused(pEventListener);
-        pEventListener.changeFocus(true);
+        //Vivecraft - dont autofocus text fields, it forces open the keyboard
+        if (Minecraft.getInstance().vrSettings.seated)
+        {
+            this.setFocused(pEventListener);
+            pEventListener.changeFocus(true);
+        }
+        //
     }
 
 default void magicalSpecialHackyFocus(@Nullable GuiEventListener pEventListener)
@@ -113,23 +119,23 @@
         else
         {
             List <? extends GuiEventListener > list = this.children();
-            int j = list.indexOf(guieventlistener);
-            int i;
+            int i = list.indexOf(guieventlistener);
+            int j;
 
-            if (flag && j >= 0)
+            if (flag && i >= 0)
             {
-                i = j + (pFocus ? 1 : 0);
+                j = i + (pFocus ? 1 : 0);
             }
             else if (pFocus)
             {
-                i = 0;
+                j = 0;
             }
             else
             {
-                i = list.size();
+                j = list.size();
             }
 
-            ListIterator <? extends GuiEventListener > listiterator = list.listIterator(i);
+            ListIterator <? extends GuiEventListener > listiterator = list.listIterator(j);
             BooleanSupplier booleansupplier = pFocus ? listiterator::hasNext : listiterator::hasPrevious;
             Supplier <? extends GuiEventListener > supplier = pFocus ? listiterator::next : listiterator::previous;
 
