--- a/net/minecraft/client/renderer/culling/Frustum.java
+++ b/net/minecraft/client/renderer/culling/Frustum.java
@@ -231,13 +231,13 @@
 
 	@Override
 	public void setCameraPosition(double var1, double var3, double var5) {
-		// TODO Auto-generated method stub
-		
+		this.camX = var1;
+		this.camY = var3;
+		this.camZ = var5;		
 	}
 
 	@Override
 	public boolean isBoundingBoxInFrustum(AABB var1) {
-		// TODO Auto-generated method stub
-		return false;
+		return this.cubeInFrustum(var1.minX, var1.minY, var1.minZ, var1.maxX, var1.maxY, var1.maxZ);
 	}
 }
