--- a/net/minecraft/client/resources/model/ModelBakery.java
+++ b/net/minecraft/client/resources/model/ModelBakery.java
@@ -36,6 +36,7 @@
 import javax.annotation.Nullable;
 import net.minecraft.Util;
 import net.minecraft.client.color.block.BlockColors;
+import net.minecraft.client.renderer.GameRenderer;
 import net.minecraft.client.renderer.RenderType;
 import net.minecraft.client.renderer.Sheets;
 import net.minecraft.client.renderer.block.BlockModelShaper;
@@ -73,6 +74,8 @@
 import org.apache.commons.lang3.tuple.Triple;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.vivecraft.gameplay.trackers.CameraTracker;
+import org.vivecraft.gameplay.trackers.TelescopeTracker;
 
 public class ModelBakery
 {
@@ -212,7 +215,13 @@
                 this.loadTopLevel(BlockModelShaper.stateToModelLocation(stateIn));
             });
         }
-
+        //Vivecraft
+        this.loadTopLevel(TelescopeTracker.scopeModel);
+        this.loadTopLevel(GameRenderer.thirdPersonCameraModel);
+        this.loadTopLevel(GameRenderer.thirdPersonCameraDisplayModel);
+        this.loadTopLevel(CameraTracker.cameraModel);
+        this.loadTopLevel(CameraTracker.cameraDisplayModel);
+        //
         profilerIn.popPush("items");
 
         for (ResourceLocation resourcelocation : Registry.ITEM.keySet())
