--- a/com/mojang/blaze3d/platform/InputConstants.java
+++ b/com/mojang/blaze3d/platform/InputConstants.java
@@ -12,6 +12,7 @@
 import java.util.OptionalInt;
 import java.util.function.BiFunction;
 import javax.annotation.Nullable;
+import net.minecraft.client.Minecraft;
 import net.minecraft.locale.Language;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.TextComponent;
@@ -24,6 +25,7 @@
 import org.lwjgl.glfw.GLFWKeyCallbackI;
 import org.lwjgl.glfw.GLFWMouseButtonCallbackI;
 import org.lwjgl.glfw.GLFWScrollCallbackI;
+import org.vivecraft.provider.InputSimulator;
 
 public class InputConstants
 {
@@ -185,9 +187,12 @@
 
     public static boolean isKeyDown(long pWindow, int p_84832_)
     {
-        return GLFW.glfwGetKey(pWindow, p_84832_) == 1;
+        return GLFW.glfwGetKey(pWindow, p_84832_) == 1 || InputSimulator.isKeyDown(p_84832_);
+    }
+    public static boolean isKeyDown(int p_216506_2_)
+    {
+        return GLFW.glfwGetKey(Minecraft.getInstance().getWindow().getWindow(), p_216506_2_) == 1 || InputSimulator.isKeyDown(p_216506_2_);
     }
-
     public static void setupKeyboardCallbacks(long pWindow, GLFWKeyCallbackI p_84846_, GLFWCharModsCallbackI pKeyCallback)
     {
         GLFW.glfwSetKeyCallback(pWindow, p_84846_);
@@ -239,14 +244,14 @@
         {
             methodhandle = lookup.findStatic(GLFW.class, "glfwRawMouseMotionSupported", methodtype);
             MethodHandle methodhandle1 = lookup.findStaticGetter(GLFW.class, "GLFW_RAW_MOUSE_MOTION", Integer.TYPE);
-            i = (int)methodhandle1.invokeExact();
+            i = (int) methodhandle1.invokeExact();
         }
-        catch (NoSuchFieldException | NoSuchMethodException nosuchmethodexception)
+        catch (NoSuchMethodException | NoSuchFieldException nosuchfieldexception)
         {
         }
-        catch (Throwable throwable)
+        catch (Throwable throwable1)
         {
-            throw new RuntimeException(throwable);
+            throw new RuntimeException(throwable1);
         }
 
         GLFW_RAW_MOUSE_MOTION_SUPPORTED = methodhandle;
@@ -336,16 +341,16 @@
 
     public static enum Type
     {
-        KEYSYM("key.keyboard", (p_84914_, p_84915_) -> {
-            String s = GLFW.glfwGetKeyName(p_84914_, -1);
-            return (Component)(s != null ? new TextComponent(s) : new TranslatableComponent(p_84915_));
+        KEYSYM("key.keyboard", (p_237528_0_, p_237528_1_) -> {
+            String s = GLFW.glfwGetKeyName(p_237528_0_, -1);
+            return (Component)(s != null ? new TextComponent(s) : new TranslatableComponent(p_237528_1_));
         }),
-        SCANCODE("scancode", (p_84911_, p_84912_) -> {
-            String s = GLFW.glfwGetKeyName(-1, p_84911_);
-            return (Component)(s != null ? new TextComponent(s) : new TranslatableComponent(p_84912_));
+        SCANCODE("scancode", (p_237527_0_, p_237527_1_) -> {
+            String s = GLFW.glfwGetKeyName(-1, p_237527_0_);
+            return (Component)(s != null ? new TextComponent(s) : new TranslatableComponent(p_237527_1_));
         }),
-        MOUSE("key.mouse", (p_84904_, p_84905_) -> {
-            return Language.getInstance().has(p_84905_) ? new TranslatableComponent(p_84905_) : new TranslatableComponent("key.mouse", p_84904_ + 1);
+        MOUSE("key.mouse", (p_237524_0_, p_237524_1_) -> {
+            return Language.getInstance().has(p_237524_1_) ? new TranslatableComponent(p_237524_1_) : new TranslatableComponent("key.mouse", p_237524_0_ + 1);
         });
 
         private final Int2ObjectMap<InputConstants.Key> map = new Int2ObjectOpenHashMap<>();
@@ -366,17 +371,17 @@
 
         public InputConstants.Key getOrCreate(int pKeyCode)
         {
-            return this.map.computeIfAbsent(pKeyCode, (p_84907_) ->
+            return this.map.computeIfAbsent(pKeyCode, (p_237525_1_) ->
             {
-                int i = p_84907_;
+                int i = p_237525_1_;
 
                 if (this == MOUSE)
                 {
-                    i = p_84907_ + 1;
+                    i = p_237525_1_ + 1;
                 }
 
                 String s = this.defaultPrefix + "." + i;
-                return new InputConstants.Key(s, this, p_84907_);
+                return new InputConstants.Key(s, this, p_237525_1_);
             });
         }
 
