--- a/net/minecraft/world/entity/projectile/ThrowableProjectile.java
+++ b/net/minecraft/world/entity/projectile/ThrowableProjectile.java
@@ -4,6 +4,7 @@
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.world.entity.EntityType;
 import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.entity.BlockEntity;
@@ -12,6 +13,9 @@
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
+import net.optifine.reflect.Reflector;
+import org.vivecraft.api.NetworkHelper;
+import org.vivecraft.api.ServerVivePlayer;
 
 public abstract class ThrowableProjectile extends Projectile
 {
@@ -30,6 +34,14 @@
     {
         this(p_37462_, p_37463_.getX(), p_37463_.getEyeY() - (double)0.1F, p_37463_.getZ(), p_37464_);
         this.setOwner(p_37463_);
+        ServerVivePlayer serverviveplayer = NetworkHelper.vivePlayers.get(p_37463_.getUUID());
+
+        if (serverviveplayer != null && serverviveplayer.isVR())
+        {
+            Vec3 vec3 = serverviveplayer.getControllerPos(serverviveplayer.activeHand, (Player)p_37463_);
+            Vec3 vec31 = serverviveplayer.getControllerDir(serverviveplayer.activeHand).scale((double)0.6F);
+            this.setPos(vec3.x + vec31.x, vec3.y + vec31.y, vec3.z + vec31.z);
+        }
     }
 
     public boolean shouldRenderAtSqrDistance(double pDistance)
@@ -76,7 +88,15 @@
 
         if (hitresult.getType() != HitResult.Type.MISS && !flag)
         {
-            this.onHit(hitresult);
+            if (Reflector.ForgeEventFactory_onProjectileImpactT.exists())
+            {
+                flag = Reflector.callBoolean(Reflector.ForgeEventFactory_onProjectileImpactT, this, hitresult);
+            }
+
+            if (!flag)
+            {
+                this.onHit(hitresult);
+            }
         }
 
         this.checkInsideBlocks();
