--- a/net/minecraft/client/main/Main.java
+++ b/net/minecraft/client/main/Main.java
@@ -44,7 +44,10 @@
 public class Main
 {
     static final Logger LOGGER = LogManager.getLogger();
-
+    public static boolean kiosk;
+    public static boolean viewonly;
+    public static boolean katvr;
+    public static boolean infinadeck;
     @DontObfuscate
     public static void main(String[] pArgs)
     {
@@ -54,6 +57,11 @@
         optionparser.accepts("demo");
         optionparser.accepts("disableMultiplayer");
         optionparser.accepts("disableChat");
+        optionparser.accepts("demo");
+        optionparser.accepts("kiosk");
+        optionparser.accepts("viewonly");
+        optionparser.accepts("katvr");
+        optionparser.accepts("infinadeck");
         optionparser.accepts("fullscreen");
         optionparser.accepts("checkGlErrors");
         OptionSpec<Void> optionspec = optionparser.accepts("jfrProfile");
@@ -72,8 +80,8 @@
         OptionSpec<String> optionspec13 = optionparser.accepts("clientId").withOptionalArg().defaultsTo("");
         OptionSpec<String> optionspec14 = optionparser.accepts("accessToken").withRequiredArg().required();
         OptionSpec<String> optionspec15 = optionparser.accepts("version").withRequiredArg().required();
-        OptionSpec<Integer> optionspec16 = optionparser.accepts("width").withRequiredArg().ofType(Integer.class).defaultsTo(854);
-        OptionSpec<Integer> optionspec17 = optionparser.accepts("height").withRequiredArg().ofType(Integer.class).defaultsTo(480);
+        OptionSpec<Integer> optionspec16 = optionparser.accepts("width").withRequiredArg().ofType(Integer.class).defaultsTo(1280);
+        OptionSpec<Integer> optionspec17 = optionparser.accepts("height").withRequiredArg().ofType(Integer.class).defaultsTo(720);
         OptionSpec<Integer> optionspec18 = optionparser.accepts("fullscreenWidth").withRequiredArg().ofType(Integer.class);
         OptionSpec<Integer> optionspec19 = optionparser.accepts("fullscreenHeight").withRequiredArg().ofType(Integer.class);
         OptionSpec<String> optionspec20 = optionparser.accepts("userProperties").withRequiredArg().defaultsTo("{}");
@@ -124,6 +132,25 @@
         OptionalInt optionalint1 = ofNullable(parseArgument(optionset, optionspec19));
         boolean flag = optionset.has("fullscreen");
         boolean flag1 = optionset.has("demo");
+        kiosk = optionset.has("kiosk");
+
+        if (kiosk)
+        {
+            System.out.println("Setting kiosk");
+        }
+
+        if (kiosk)
+        {
+            viewonly = optionset.has("viewonly");
+
+            if (viewonly)
+            {
+                System.out.println("Setting viewonly");
+            }
+        }
+
+        katvr = optionset.has("katvr");
+        infinadeck = optionset.has("infinadeck");
         boolean flag2 = optionset.has("disableMultiplayer");
         boolean flag3 = optionset.has("disableChat");
         String s3 = parseArgument(optionset, optionspec15);
@@ -196,6 +223,7 @@
         }
         catch (Throwable throwable1)
         {
+            LOGGER.warn("Failed to something: ", throwable1);
             CrashReport crashreport = CrashReport.forThrowable(throwable1, "Initializing game");
             CrashReportCategory crashreportcategory = crashreport.addCategory("Initialization");
             NativeModuleLister.addCrashSection(crashreportcategory);
@@ -255,9 +283,9 @@
                 thread1.join();
             }
         }
-        catch (InterruptedException interruptedexception)
+        catch (InterruptedException interruptedexception1)
         {
-            LOGGER.error("Exception during client thread shutdown", (Throwable)interruptedexception);
+            LOGGER.error("Exception during client thread shutdown", (Throwable)interruptedexception1);
         }
         finally
         {
@@ -307,6 +335,6 @@
     static
     {
         Util.preInitLog4j();
-        System.setProperty("java.awt.headless", "true");
+        //Vivecraft we do this for some reason. System.setProperty("java.awt.headless", "true");
     }
 }
