--- a/net/minecraft/client/renderer/ItemInHandRenderer.java
+++ b/net/minecraft/client/renderer/ItemInHandRenderer.java
@@ -6,8 +6,14 @@
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.blaze3d.vertex.VertexConsumer;
 import com.mojang.math.Matrix4f;
+import com.mojang.math.Quaternion;
 import com.mojang.math.Vector3f;
+
 import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import net.minecraft.Util;
 import net.minecraft.client.Minecraft;
 import net.minecraft.client.player.AbstractClientPlayer;
 import net.minecraft.client.player.LocalPlayer;
@@ -16,21 +22,48 @@
 import net.minecraft.client.renderer.entity.ItemRenderer;
 import net.minecraft.client.renderer.entity.player.PlayerRenderer;
 import net.minecraft.client.renderer.texture.OverlayTexture;
+import net.minecraft.client.resources.model.BakedModel;
+import net.minecraft.core.BlockPos;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.util.Mth;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.entity.HumanoidArm;
 import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.item.ArrowItem;
+import net.minecraft.world.item.BlockItem;
+import net.minecraft.world.item.CompassItem;
 import net.minecraft.world.item.CrossbowItem;
+import net.minecraft.world.item.FishingRodItem;
+import net.minecraft.world.item.FoodOnAStickItem;
+import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
 import net.minecraft.world.item.MapItem;
+import net.minecraft.world.item.ShieldItem;
+import net.minecraft.world.item.SwordItem;
+import net.minecraft.world.item.TridentItem;
+import net.minecraft.world.item.UseAnim;
+import net.minecraft.world.item.enchantment.EnchantmentHelper;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.TorchBlock;
+import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.saveddata.maps.MapItemSavedData;
+import net.minecraft.world.phys.AABB;
+import net.minecraft.world.phys.Vec3;
 import net.optifine.Config;
 import net.optifine.CustomItems;
 import net.optifine.reflect.Reflector;
 import net.optifine.shaders.Shaders;
-
+import org.apache.commons.lang3.tuple.Triple;
+import org.vivecraft.gameplay.trackers.BowTracker;
+import org.vivecraft.gameplay.trackers.SwingTracker;
+import org.vivecraft.gameplay.trackers.TelescopeTracker;
+import org.vivecraft.provider.ControllerType;
+import org.vivecraft.render.RenderPass;
+import org.vivecraft.render.VRArmRenderer;
+import org.vivecraft.render.VRFirstPersonArmSwing;
+import org.vivecraft.render.VivecraftItemRendering;
+//TODO: this needs a lot of work
 public class ItemInHandRenderer
 {
     private static final RenderType MAP_BACKGROUND = RenderType.text(new ResourceLocation("textures/map/map_background.png"));
@@ -115,6 +148,8 @@
     private final EntityRenderDispatcher entityRenderDispatcher;
     private final ItemRenderer itemRenderer;
     private static boolean renderItemHand = false;
+    public VRFirstPersonArmSwing swingType = VRFirstPersonArmSwing.Attack;
+    public float xdist = 0.0F;
 
     public ItemInHandRenderer(Minecraft p_109310_)
     {
@@ -123,6 +158,7 @@
         this.itemRenderer = p_109310_.getItemRenderer();
     }
 
+    
     public void renderItem(LivingEntity pLivingEntity, ItemStack pItemStack, ItemTransforms.TransformType pTransformType, boolean pLeftHand, PoseStack pMatrixStack, MultiBufferSource pBuffer, int pCombinedLight)
     {
         CustomItems.setRenderOffHand(pLeftHand);
@@ -245,33 +281,49 @@
     {
         boolean flag = pSide != HumanoidArm.LEFT;
         float f = flag ? 1.0F : -1.0F;
-        float f1 = Mth.sqrt(pSwingProgress);
-        float f2 = -0.3F * Mth.sin(f1 * (float)Math.PI);
-        float f3 = 0.4F * Mth.sin(f1 * ((float)Math.PI * 2F));
-        float f4 = -0.4F * Mth.sin(pSwingProgress * (float)Math.PI);
-        pMatrixStack.translate((double)(f * (f2 + 0.64000005F)), (double)(f3 + -0.6F + pEquippedProgress * -0.6F), (double)(f4 + -0.71999997F));
-        pMatrixStack.mulPose(Vector3f.YP.rotationDegrees(f * 45.0F));
-        float f5 = Mth.sin(pSwingProgress * pSwingProgress * (float)Math.PI);
-        float f6 = Mth.sin(f1 * (float)Math.PI);
-        pMatrixStack.mulPose(Vector3f.YP.rotationDegrees(f * f6 * 70.0F));
-        pMatrixStack.mulPose(Vector3f.ZP.rotationDegrees(f * f5 * -20.0F));
+//        float f1 = Mth.sqrt(pSwingProgress);
+//        float f2 = -0.3F * Mth.sin(f1 * (float)Math.PI);
+//        float f3 = 0.4F * Mth.sin(f1 * ((float)Math.PI * 2F));
+//        float f4 = -0.4F * Mth.sin(pSwingProgress * (float)Math.PI);
+//        pMatrixStack.translate((double)(f * (f2 + 0.64000005F)), (double)(f3 + -0.6F + pEquippedProgress * -0.6F), (double)(f4 + -0.71999997F));
+//        pMatrixStack.mulPose(Vector3f.YP.rotationDegrees(f * 45.0F));
+//        float f5 = Mth.sin(pSwingProgress * pSwingProgress * (float)Math.PI);
+//        float f6 = Mth.sin(f1 * (float)Math.PI);
+//        pMatrixStack.mulPose(Vector3f.YP.rotationDegrees(f * f6 * 70.0F));
+//        pMatrixStack.mulPose(Vector3f.ZP.rotationDegrees(f * f5 * -20.0F));
         AbstractClientPlayer abstractclientplayer = this.minecraft.player;
         RenderSystem.setShaderTexture(0, abstractclientplayer.getSkinTextureLocation());
-        pMatrixStack.translate((double)(f * -1.0F), (double)3.6F, 3.5D);
-        pMatrixStack.mulPose(Vector3f.ZP.rotationDegrees(f * 120.0F));
-        pMatrixStack.mulPose(Vector3f.XP.rotationDegrees(200.0F));
-        pMatrixStack.mulPose(Vector3f.YP.rotationDegrees(f * -135.0F));
-        pMatrixStack.translate((double)(f * 5.6F), 0.0D, 0.0D);
-        PlayerRenderer playerrenderer = (PlayerRenderer)this.entityRenderDispatcher.<AbstractClientPlayer>getRenderer(abstractclientplayer);
-
-        if (flag)
-        {
-            playerrenderer.renderRightHand(pMatrixStack, pBuffer, pCombinedLight, abstractclientplayer);
-        }
-        else
-        {
-            playerrenderer.renderLeftHand(pMatrixStack, pBuffer, pCombinedLight, abstractclientplayer);
-        }
+        VRArmRenderer vrarmrenderer = entityRenderDispatcher.armSkinMap.get(abstractclientplayer.getModelName());
+        pMatrixStack.pushPose();
+	
+	        if (abstractclientplayer.swingingArm == InteractionHand.MAIN_HAND && flag)
+	        {
+	            this.transformFirstPersonVR(pMatrixStack, pSide, pSwingProgress);
+	        }
+	
+	        if (abstractclientplayer.swingingArm == InteractionHand.OFF_HAND && !flag)
+	        {
+	            this.transformFirstPersonVR(pMatrixStack, pSide, pSwingProgress);
+	        }
+	//        pMatrixStack.translate((double)(f * -1.0F), (double)3.6F, 3.5D);
+	//        pMatrixStack.mulPose(Vector3f.ZP.rotationDegrees(f * 120.0F));
+	//        pMatrixStack.mulPose(Vector3f.XP.rotationDegrees(200.0F));
+	//        pMatrixStack.mulPose(Vector3f.YP.rotationDegrees(f * -135.0F));
+	//        pMatrixStack.translate((double)(f * 5.6F), 0.0D, 0.0D);
+	        pMatrixStack.scale(0.4f, 0.4F, 0.4F);
+	        boolean slim = abstractclientplayer.getSkinTextureLocation().getPath().equals("slim");
+	        pMatrixStack.translate(slim ? 0.345F * -f : 0.375F * -f, 0, slim ? 0.785F : 0.75F);
+	        pMatrixStack.mulPose(Vector3f.XP.rotationDegrees(-90));
+	        pMatrixStack.mulPose(Vector3f.YP.rotationDegrees(180));	
+	        if (flag)
+	        {
+	            vrarmrenderer.renderRightHand(pMatrixStack, pBuffer, pCombinedLight, abstractclientplayer);
+	        }
+	        else
+	        {
+	            vrarmrenderer.renderLeftHand(pMatrixStack, pBuffer, pCombinedLight, abstractclientplayer);
+	        }
+       pMatrixStack.popPose();
     }
 
     private void applyEatTransform(PoseStack pMatrixStack, float pPartialTicks, HumanoidArm pHand, ItemStack pStack)
@@ -304,6 +356,49 @@
         pMatrixStack.mulPose(Vector3f.YP.rotationDegrees((float)i * -45.0F));
     }
 
+    private void transformFirstPersonVR(PoseStack matrixStackIn, HumanoidArm hand, float swingProgress)
+    {
+        if (swingProgress != 0.0F)
+        {
+            switch (this.swingType)
+            {
+                case Attack:
+                    float f2 = Mth.sin((float)((double)(swingProgress * 3.0F) * Math.PI));
+
+                    if ((double)swingProgress > 0.5D)
+                    {
+                        f2 = Mth.sin((float)((double)swingProgress * Math.PI + Math.PI));
+                    }
+
+                    matrixStackIn.translate(0.0D, 0.0D, (double)0.2F);
+                    matrixStackIn.mulPose(Vector3f.XP.rotationDegrees(f2 * 30.0F));
+                    matrixStackIn.translate(0.0D, 0.0D, (double) - 0.2F);
+                    break;
+
+                case Interact:
+                    float f1 = Mth.sin((float)((double)(swingProgress * 3.0F) * Math.PI));
+
+                    if ((double)swingProgress > 0.5D)
+                    {
+                        f1 = Mth.sin((float)((double)swingProgress * Math.PI + Math.PI));
+                    }
+
+                    matrixStackIn.mulPose(Vector3f.ZP.rotationDegrees((float)(hand == HumanoidArm.RIGHT ? -1 : 1) * f1 * 45.0F));
+                    break;
+
+                case Use:
+                    float f = Mth.sin((float)((double)(swingProgress * 2.0F) * Math.PI));
+
+                    if ((double)swingProgress > 0.25D)
+                    {
+                        f = Mth.sin((float)((double)(swingProgress / 2.0F) * Math.PI + Math.PI));
+                    }
+
+                    matrixStackIn.translate(0.0D, 0.0D, (double)(-(1.0F + f) * 0.1F));
+            }
+        }
+    }
+
     private void applyItemArmTransform(PoseStack pMatrixStack, HumanoidArm pHand, float pEquippedProg)
     {
         int i = pHand == HumanoidArm.RIGHT ? 1 : -1;
@@ -388,188 +483,481 @@
         return p_172913_.is(Items.CROSSBOW) && CrossbowItem.isCharged(p_172913_);
     }
 
-    private void renderArmWithItem(AbstractClientPlayer pPlayer, float pPartialTicks, float pPitch, InteractionHand pHand, float pSwingProgress, ItemStack pStack, float pEquippedProgress, PoseStack pMatrixStack, MultiBufferSource pBuffer, int pCombinedLight)
+    void renderArmWithItem(AbstractClientPlayer pPlayer, float pPartialTicks, float pPitch, InteractionHand pHand, float pSwingProgress, ItemStack pStack, float pEquippedProgress, PoseStack pMatrixStack, MultiBufferSource pBuffer, int pCombinedLight)
     {
         if (!Config.isShaders() || !Shaders.isSkipRenderHand(pHand))
         {
-            if (!pPlayer.isScoping())
+            boolean flag = pHand == InteractionHand.MAIN_HAND;
+            HumanoidArm humanoidarm = flag ? pPlayer.getMainArm() : pPlayer.getMainArm().getOpposite();
+            pEquippedProgress = this.getEquipProgress(pHand, pPartialTicks);
+            pMatrixStack.pushPose();
+            boolean flag1 = true;
+
+            if (this.minecraft.currentPass == RenderPass.THIRD && !this.minecraft.vrSettings.mixedRealityRenderHands)
             {
-                boolean flag = pHand == InteractionHand.MAIN_HAND;
-                HumanoidArm humanoidarm = flag ? pPlayer.getMainArm() : pPlayer.getMainArm().getOpposite();
-                pMatrixStack.pushPose();
+                flag1 = false;
+            }
+
+            if (this.minecraft.currentPass == RenderPass.CAMERA)
+            {
+                flag1 = false;
+            }
+
+            if (BowTracker.isBow(pStack) && this.minecraft.bowTracker.isActive((LocalPlayer)pPlayer))
+            {
+                flag1 = false;
+            }
+
+            if (TelescopeTracker.isTelescope(pStack) && (pHand == InteractionHand.OFF_HAND && this.minecraft.currentPass == RenderPass.SCOPEL || pHand == InteractionHand.MAIN_HAND && this.minecraft.currentPass == RenderPass.SCOPER))
+            {
+                flag1 = false;
+            }
+
+            if (flag1 && !pPlayer.isInvisible())
+            {
+                this.renderPlayerArm(pMatrixStack, pBuffer, pCombinedLight, pEquippedProgress, pSwingProgress, humanoidarm);
+            }
 
-                if (pStack.isEmpty())
+            if (!pStack.isEmpty())
+            {
+                Item item = pStack.getItem();
+                boolean flag2 = false;
+                VivecraftItemRendering vivecraftitemrendering = VivecraftItemRendering.Item;
+
+                if (pStack.getUseAnimation() != UseAnim.EAT && pStack.getUseAnimation() != UseAnim.DRINK)
                 {
-                    if (flag && !pPlayer.isInvisible())
+                    if (item instanceof BlockItem)
+                    {
+                        Block block = ((BlockItem)item).getBlock();
+
+                        if (block instanceof TorchBlock)
+                        {
+                            vivecraftitemrendering = VivecraftItemRendering.Block_Stick;
+                        }
+                        else
+                        {
+                            BakedModel bakedmodel = this.itemRenderer.getModel(pStack, this.minecraft.level, this.minecraft.player, 0);
+
+                            if (bakedmodel.isGui3d())
+                            {
+                                vivecraftitemrendering = VivecraftItemRendering.Block_3D;
+                            }
+                            else
+                            {
+                                vivecraftitemrendering = VivecraftItemRendering.Block_Item;
+                            }
+                        }
+                    }
+                    else if (item instanceof MapItem)
                     {
-                        this.renderPlayerArm(pMatrixStack, pBuffer, pCombinedLight, pEquippedProgress, pSwingProgress, humanoidarm);
+                        vivecraftitemrendering = VivecraftItemRendering.Map;
                     }
-                }
-                else if (pStack.getItem() instanceof MapItem)
-                {
-                    if (flag && this.offHandItem.isEmpty())
+                    else if (pStack.getUseAnimation() == UseAnim.BOW)
+                    {
+                        vivecraftitemrendering = VivecraftItemRendering.Bow_Seated;
+
+                        if (this.minecraft.bowTracker.isActive((LocalPlayer)pPlayer))
+                        {
+                            if (this.minecraft.bowTracker.isDrawing)
+                            {
+                                vivecraftitemrendering = VivecraftItemRendering.Bow_Roomscale_Drawing;
+                            }
+                            else
+                            {
+                                vivecraftitemrendering = VivecraftItemRendering.Bow_Roomscale;
+                            }
+                        }
+                    }
+                    else if (item instanceof SwordItem)
+                    {
+                        vivecraftitemrendering = VivecraftItemRendering.Sword;
+                    }
+                    else if (item instanceof ShieldItem)
                     {
-                        this.renderTwoHandedMap(pMatrixStack, pBuffer, pCombinedLight, pPitch, pEquippedProgress, pSwingProgress);
+                        vivecraftitemrendering = VivecraftItemRendering.Shield;
+                    }
+                    else if (item instanceof TridentItem)
+                    {
+                        vivecraftitemrendering = VivecraftItemRendering.Spear;
+                    }
+                    else if (item instanceof CrossbowItem)
+                    {
+                        vivecraftitemrendering = VivecraftItemRendering.Crossbow;
+                    }
+                    else if (!(item instanceof CompassItem) && item != Items.CLOCK)
+                    {
+                        if (SwingTracker.isTool(item))
+                        {
+                            vivecraftitemrendering = VivecraftItemRendering.Tool;
+
+                            if (item instanceof FoodOnAStickItem || item instanceof FishingRodItem)
+                            {
+                                vivecraftitemrendering = VivecraftItemRendering.Tool_Rod;
+                            }
+                        }
+                        else if (TelescopeTracker.isTelescope(pStack))
+                        {
+                            vivecraftitemrendering = VivecraftItemRendering.Telescope;
+                        }
                     }
                     else
                     {
-                        this.renderOneHandedMap(pMatrixStack, pBuffer, pCombinedLight, pEquippedProgress, humanoidarm, pSwingProgress, pStack);
+                        vivecraftitemrendering = VivecraftItemRendering.Compass;
                     }
                 }
-                else if (pStack.is(Items.CROSSBOW))
+                else
                 {
-                    boolean flag2 = CrossbowItem.isCharged(pStack);
-                    boolean flag3 = humanoidarm == HumanoidArm.RIGHT;
-                    int l = flag3 ? 1 : -1;
+                    vivecraftitemrendering = VivecraftItemRendering.Noms;
+                }
+
+                int k = flag ? 1 : -1;
+                double d8 = 0.7D;
+                double d0 = -0.05D;
+                double d1 = 0.005D;
+                double d2 = 0.0D;
+                double d3 = this.minecraft.vr.getGunAngle();
+                Quaternion quaternion = Vector3f.YP.rotationDegrees(0.0F);
+                Quaternion quaternion1 = Vector3f.YP.rotationDegrees(0.0F);
+                quaternion.mul(Vector3f.XP.rotationDegrees((float)(-110.0D + d3)));
+                pMatrixStack.pushPose();
+                boolean flag3 = false;
 
-                    if (pPlayer.isUsingItem() && pPlayer.getUseItemRemainingTicks() > 0 && pPlayer.getUsedItemHand() == pHand)
+                if (vivecraftitemrendering == VivecraftItemRendering.Bow_Seated)
+                {
+                    d1 += -0.1D;
+                    d2 += 0.1D;
+                    quaternion.mul(Vector3f.XP.rotationDegrees((float)(90.0D - d3)));
+                    d8 = (double)0.7F;
+                }
+                else if (vivecraftitemrendering == VivecraftItemRendering.Bow_Roomscale)
+                {
+                    quaternion = Vector3f.XP.rotationDegrees(0.0F);
+                    pMatrixStack.mulPose(Vector3f.XP.rotationDegrees((float)(-110.0D + d3)));
+                    d1 -= 0.25D;
+                    d2 += (double)0.025F + 0.03D * d3 / 40.0D;
+                    d0 += -0.0225D;
+                    d8 = 1.0D;
+                }
+                else if (vivecraftitemrendering == VivecraftItemRendering.Bow_Roomscale_Drawing)
+                {
+                    quaternion = Vector3f.YP.rotationDegrees(0.0F);
+                    d8 = 1.0D;
+                    int i = 0;
+
+                    if (this.minecraft.vrSettings.reverseShootingEye)
                     {
-                        this.applyItemArmTransform(pMatrixStack, humanoidarm, pEquippedProgress);
-                        pMatrixStack.translate((double)((float)l * -0.4785682F), (double) - 0.094387F, (double)0.05731531F);
-                        pMatrixStack.mulPose(Vector3f.XP.rotationDegrees(-11.935F));
-                        pMatrixStack.mulPose(Vector3f.YP.rotationDegrees((float)l * 65.3F));
-                        pMatrixStack.mulPose(Vector3f.ZP.rotationDegrees((float)l * -9.785F));
-                        float f10 = (float)pStack.getUseDuration() - ((float)this.minecraft.player.getUseItemRemainingTicks() - pPartialTicks + 1.0F);
-                        float f14 = f10 / (float)CrossbowItem.getChargeDuration(pStack);
+                        i = 1;
+                    }
 
-                        if (f14 > 1.0F)
-                        {
-                            f14 = 1.0F;
-                        }
+                    Vec3 vec3 = this.minecraft.bowTracker.getAimVector();
+                    Vec3 vec31 = new Vec3(vec3.x, vec3.y, vec3.z);
+                    Vec3 vec32 = this.minecraft.vrPlayer.vrdata_world_render.getHand(1).getCustomVector(new Vec3(0.0D, -1.0D, 0.0D));
+                    Vec3 vec33 = this.minecraft.vrPlayer.vrdata_world_render.getHand(1).getCustomVector(new Vec3(0.0D, 0.0D, -1.0D));
+                    vec31.cross(vec32);
+                    double d4 = (180D / Math.PI) * Math.acos(vec31.dot(vec32));
+                    float f = (float)Math.toDegrees(Math.asin(vec31.y / vec31.length()));
+                    float f1 = (float)Math.toDegrees(Math.atan2(vec31.x, vec31.z));
+                    Vec3 vec34 = new Vec3(0.0D, 1.0D, 0.0D);
+                    Vec3 vec35 = new Vec3(vec31.x, 0.0D, vec31.z);
+                    Vec3 vec36 = Vec3.ZERO;
+                    double d5 = vec33.dot(vec35);
 
-                        if (f14 > 0.1F)
-                        {
-                            float f17 = Mth.sin((f10 - 0.1F) * 1.3F);
-                            float f19 = f14 - 0.1F;
-                            float f20 = f17 * f19;
-                            pMatrixStack.translate((double)(f20 * 0.0F), (double)(f20 * 0.004F), (double)(f20 * 0.0F));
-                        }
+                    if (d5 != 0.0D)
+                    {
+                        vec36 = vec35.scale(d5);
+                    }
+
+                    double d6 = 0.0D;
+                    Vec3 vec37 = vec33.subtract(vec36).normalize();
+                    d6 = vec37.dot(vec34);
+                    double d7 = vec35.dot(vec37.cross(vec34));
+                    float f2;
 
-                        pMatrixStack.translate((double)(f14 * 0.0F), (double)(f14 * 0.0F), (double)(f14 * 0.04F));
-                        pMatrixStack.scale(1.0F, 1.0F, 1.0F + f14 * 0.2F);
-                        pMatrixStack.mulPose(Vector3f.YN.rotationDegrees((float)l * 45.0F));
+                    if (d7 < 0.0D)
+                    {
+                        f2 = -((float)Math.acos(d6));
                     }
                     else
                     {
-                        float f9 = -0.4F * Mth.sin(Mth.sqrt(pSwingProgress) * (float)Math.PI);
-                        float f13 = 0.2F * Mth.sin(Mth.sqrt(pSwingProgress) * ((float)Math.PI * 2F));
-                        float f16 = -0.2F * Mth.sin(pSwingProgress * (float)Math.PI);
-                        pMatrixStack.translate((double)((float)l * f9), (double)f13, (double)f16);
-                        this.applyItemArmTransform(pMatrixStack, humanoidarm, pEquippedProgress);
-                        this.applyItemArmAttackTransform(pMatrixStack, humanoidarm, pSwingProgress);
+                        f2 = (float)Math.acos(d6);
+                    }
 
-                        if (flag2 && pSwingProgress < 0.001F && flag)
-                        {
-                            pMatrixStack.translate((double)((float)l * -0.641864F), 0.0D, 0.0D);
-                            pMatrixStack.mulPose(Vector3f.YP.rotationDegrees((float)l * 10.0F));
-                        }
+                    float f3 = (float)((180D / Math.PI) * (double)f2);
+
+                    if (this.minecraft.bowTracker.isCharged())
+                    {
+                        long j = Util.getMillis() - this.minecraft.bowTracker.startDrawTime;
+                        d0 += 0.003D * Math.sin((double)j);
                     }
 
-                    this.renderItem(pPlayer, pStack, flag3 ? ItemTransforms.TransformType.FIRST_PERSON_RIGHT_HAND : ItemTransforms.TransformType.FIRST_PERSON_LEFT_HAND, !flag3, pMatrixStack, pBuffer, pCombinedLight);
+                    pMatrixStack.translate(0.0D, 0.0D, 0.1D);
+                    pMatrixStack.last().pose().multiply(this.minecraft.vrPlayer.vrdata_world_render.getController(1).getMatrix().transposed().toMCMatrix());
+                    quaternion.mul(Vector3f.YP.rotationDegrees(f1));
+                    quaternion.mul(Vector3f.XP.rotationDegrees(-f));
+                    quaternion.mul(Vector3f.ZP.rotationDegrees(-f3));
+                    quaternion.mul(Vector3f.ZP.rotationDegrees(180.0F));
+                    pMatrixStack.last().pose().multiply(quaternion);
+                    quaternion = Vector3f.YP.rotationDegrees(0.0F);
+                    quaternion.mul(Vector3f.YP.rotationDegrees(180.0F));
+                    quaternion.mul(Vector3f.XP.rotationDegrees(160.0F));
+                    d1 += 0.1225D;
+                    d0 += 0.125D;
+                    d2 += 0.16D;
                 }
-                else
+                else if (vivecraftitemrendering == VivecraftItemRendering.Crossbow)
+                {
+                    d0 += (double)0.01F;
+                    d2 += (double) - 0.02F;
+                    d1 += (double) - 0.02F;
+                    d8 = 0.5D;
+                    quaternion = Vector3f.XP.rotationDegrees(0.0F);
+                    quaternion.mul(Vector3f.YP.rotationDegrees(10.0F));
+                }
+                else if (vivecraftitemrendering == VivecraftItemRendering.Map)
+                {
+                    flag2 = true;
+                    quaternion = Vector3f.XP.rotationDegrees(-45.0F);
+                    d0 = 0.0D;
+                    d1 = 0.16D;
+                    d2 = -0.075D;
+                    d8 = 0.75D;
+                }
+                else if (vivecraftitemrendering == VivecraftItemRendering.Noms)
+                {
+                    long l = (long)this.minecraft.player.getUseItemRemainingTicks();
+                    quaternion = Vector3f.ZP.rotationDegrees(180.0F);
+                    quaternion.mul(Vector3f.XP.rotationDegrees(-135.0F));
+                    d2 = d2 + 0.006D * Math.sin((double)l);
+                    d2 = d2 + (double)0.02F;
+                    d0 += (double)0.08F;
+                    d8 = (double)0.4F;
+                }
+                else if (vivecraftitemrendering != VivecraftItemRendering.Item && vivecraftitemrendering != VivecraftItemRendering.Block_Item)
                 {
-                    boolean flag1 = humanoidarm == HumanoidArm.RIGHT;
+                    if (vivecraftitemrendering == VivecraftItemRendering.Compass)
+                    {
+                        quaternion = Vector3f.YP.rotationDegrees(90.0F);
+                        quaternion.mul(Vector3f.XP.rotationDegrees(25.0F));
+                        d8 = (double)0.4F;
+                    }
+                    else if (vivecraftitemrendering == VivecraftItemRendering.Block_3D)
+                    {
+                        d8 = (double)0.3F;
+                        d2 += (double) - 0.1F;
+                        d0 += (double)0.05F;
+                    }
+                    else if (vivecraftitemrendering == VivecraftItemRendering.Block_Stick)
+                    {
+                        quaternion = Vector3f.XP.rotationDegrees(0.0F);
+                        d1 += -0.105D + 0.06D * d3 / 40.0D;
+                        d2 += (double) - 0.1F;
+                        quaternion.mul(Vector3f.XP.rotationDegrees(-45.0F));
+                        quaternion.mul(Vector3f.XP.rotationDegrees((float)d3));
+                    }
+                    else if (vivecraftitemrendering == VivecraftItemRendering.Shield)
+                    {
+                        flag3 = !flag;
+                        d8 = (double)0.4F;
+                        d1 += (double)0.21F;
+
+                        if (flag)
+                        {
+                            d0 += (double)0.11F;
+                        }
+                        else
+                        {
+                            d0 += -0.015D;
+                        }
+
+                        d2 += (double) - 0.01F;
+                        quaternion.mul(Vector3f.XP.rotationDegrees((float)(105.0D - d3)));
+
+                        if (pPlayer.isUsingItem() && pPlayer.getUseItemRemainingTicks() > 0 && pPlayer.getUsedItemHand() == pHand)
+                        {
+                            quaternion.mul(Vector3f.ZP.rotationDegrees((float)(k * -5)));
+                            d1 += (double) - 0.13F;
+                            d0 += (double)((float)k * 0.05F);
+                            d2 += (double) - 0.1F;
+
+                            if (pPlayer.isBlocking())
+                            {
+                                quaternion.mul(Vector3f.YP.rotationDegrees((float)k * 90.0F));
+                            }
+                            else
+                            {
+                                quaternion.mul(Vector3f.YP.rotationDegrees((1.0F - pEquippedProgress) * (float)k * 90.0F));
+                            }
+                        }
 
-                    if (pPlayer.isUsingItem() && pPlayer.getUseItemRemainingTicks() > 0 && pPlayer.getUsedItemHand() == pHand)
+                        quaternion.mul(Vector3f.YP.rotationDegrees((float)k * -90.0F));
+                    }
+                    else if (vivecraftitemrendering == VivecraftItemRendering.Spear)
                     {
-                        int k = flag1 ? 1 : -1;
+                        quaternion = Vector3f.XP.rotationDegrees(0.0F);
+                        d0 += (double) - 0.135F;
+                        d2 = d2 + (double)0.575F;
+                        d8 = (double)0.6F;
+                        float f4 = 0.0F;
+                        boolean flag5 = false;
+                        int i1 = 0;
 
-                        switch (pStack.getUseAnimation())
+                        if (pPlayer.isUsingItem() && pPlayer.getUseItemRemainingTicks() > 0 && pPlayer.getUsedItemHand() == pHand)
                         {
-                            case NONE:
-                                this.applyItemArmTransform(pMatrixStack, humanoidarm, pEquippedProgress);
-                                break;
-
-                            case EAT:
-                            case DRINK:
-                                this.applyEatTransform(pMatrixStack, pPartialTicks, humanoidarm, pStack);
-                                this.applyItemArmTransform(pMatrixStack, humanoidarm, pEquippedProgress);
-                                break;
-
-                            case BLOCK:
-                                this.applyItemArmTransform(pMatrixStack, humanoidarm, pEquippedProgress);
-                                break;
-
-                            case BOW:
-                                this.applyItemArmTransform(pMatrixStack, humanoidarm, pEquippedProgress);
-                                pMatrixStack.translate((double)((float)k * -0.2785682F), (double)0.18344387F, (double)0.15731531F);
-                                pMatrixStack.mulPose(Vector3f.XP.rotationDegrees(-13.935F));
-                                pMatrixStack.mulPose(Vector3f.YP.rotationDegrees((float)k * 35.3F));
-                                pMatrixStack.mulPose(Vector3f.ZP.rotationDegrees((float)k * -9.785F));
-                                float f7 = (float)pStack.getUseDuration() - ((float)this.minecraft.player.getUseItemRemainingTicks() - pPartialTicks + 1.0F);
-                                float f8 = f7 / 20.0F;
-                                f8 = (f8 * f8 + f8 * 2.0F) / 3.0F;
+                            flag5 = true;
+                            i1 = EnchantmentHelper.getRiptide(pStack);
 
-                                if (f8 > 1.0F)
-                                {
-                                    f8 = 1.0F;
-                                }
+                            if (i1 <= 0 || i1 > 0 && pPlayer.isInWaterOrRain())
+                            {
+                                f4 = (float)pStack.getUseDuration() - ((float)this.minecraft.player.getUseItemRemainingTicks() - pPartialTicks + 1.0F);
 
-                                if (f8 > 0.1F)
+                                if (f4 > 10.0F)
                                 {
-                                    float f12 = Mth.sin((f7 - 0.1F) * 1.3F);
-                                    float f15 = f8 - 0.1F;
-                                    float f18 = f12 * f15;
-                                    pMatrixStack.translate((double)(f18 * 0.0F), (double)(f18 * 0.004F), (double)(f18 * 0.0F));
-                                }
+                                    f4 = 10.0F;
 
-                                pMatrixStack.translate((double)(f8 * 0.0F), (double)(f8 * 0.0F), (double)(f8 * 0.04F));
-                                pMatrixStack.scale(1.0F, 1.0F, 1.0F + f8 * 0.2F);
-                                pMatrixStack.mulPose(Vector3f.YN.rotationDegrees((float)k * 45.0F));
-                                break;
-
-                            case SPEAR:
-                                this.applyItemArmTransform(pMatrixStack, humanoidarm, pEquippedProgress);
-                                pMatrixStack.translate((double)((float)k * -0.5F), (double)0.7F, (double)0.1F);
-                                pMatrixStack.mulPose(Vector3f.XP.rotationDegrees(-55.0F));
-                                pMatrixStack.mulPose(Vector3f.YP.rotationDegrees((float)k * 35.3F));
-                                pMatrixStack.mulPose(Vector3f.ZP.rotationDegrees((float)k * -9.785F));
-                                float f11 = (float)pStack.getUseDuration() - ((float)this.minecraft.player.getUseItemRemainingTicks() - pPartialTicks + 1.0F);
-                                float f2 = f11 / 10.0F;
+                                    if (i1 > 0 && pPlayer.isInWaterOrRain())
+                                    {
+                                        pMatrixStack.mulPose(Vector3f.ZP.rotationDegrees((float)(-this.minecraft.tickCounter * 10 * i1 % 360) - pPartialTicks * 10.0F * (float)i1));
+                                    }
+
+                                    if (this.minecraft.frameIndex % 4L == 0L)
+                                    {
+                                        this.minecraft.vr.triggerHapticPulse(flag ? 0 : 1, 200);
+                                    }
 
-                                if (f2 > 1.0F)
-                                {
-                                    f2 = 1.0F;
+                                    long j1 = Util.getMillis() - this.minecraft.bowTracker.startDrawTime;
+                                    d0 += 0.003D * Math.sin((double)j1);
                                 }
+                            }
+                        }
+
+                        if (pPlayer.isAutoSpinAttack())
+                        {
+                            i1 = 5;
+                            d2 += (double) - 0.15F;
+                            pMatrixStack.mulPose(Vector3f.ZP.rotationDegrees((float)(-this.minecraft.tickCounter * 10 * i1 % 360) - pPartialTicks * 10.0F * (float)i1));
+                            flag5 = true;
+                        }
+
+                        if (!flag5)
+                        {
+                            d1 += 0.0D + 0.2D * d3 / 40.0D;
+                            quaternion.mul(Vector3f.XP.rotationDegrees((float)d3));
+                        }
+
+                        quaternion.mul(Vector3f.XP.rotationDegrees(-65.0F));
+                        d2 = d2 + (double)(-0.75F + f4 / 10.0F * 0.25F);
+                    }
+                    else if (vivecraftitemrendering != VivecraftItemRendering.Sword)
+                    {
+                        if (vivecraftitemrendering == VivecraftItemRendering.Tool_Rod)
+                        {
+                            d2 += (double) - 0.15F;
+                            d1 += -0.02D + d3 / 40.0D * 0.1D;
+                            d0 += (double)0.05F;
+                            quaternion.mul(Vector3f.XP.rotationDegrees(40.0F));
+                            d8 = (double)0.8F;
+                        }
+                        else if (vivecraftitemrendering == VivecraftItemRendering.Tool)
+                        {
+                            boolean flag4 = this.minecraft.climbTracker.isClaws(pStack) && this.minecraft.climbTracker.isClimbeyClimb();
+
+                            if (flag4)
+                            {
+                                quaternion.mul(Vector3f.XP.rotationDegrees((float)(-d3)));
+                                d8 = (double)0.3F;
+                                d2 += (double)0.075F;
+                                d1 += (double)0.02F;
+                                d0 += (double)0.03F;
 
-                                if (f2 > 0.1F)
+                                if (this.minecraft.vr.keyClimbeyGrab.isDown(ControllerType.RIGHT) && flag || this.minecraft.vr.keyClimbeyGrab.isDown(ControllerType.LEFT) && !flag)
                                 {
-                                    float f3 = Mth.sin((f11 - 0.1F) * 1.3F);
-                                    float f4 = f2 - 0.1F;
-                                    float f5 = f3 * f4;
-                                    pMatrixStack.translate((double)(f5 * 0.0F), (double)(f5 * 0.004F), (double)(f5 * 0.0F));
+                                    d2 += (double) - 0.2F;
                                 }
+                            }
 
-                                pMatrixStack.translate(0.0D, 0.0D, (double)(f2 * 0.2F));
-                                pMatrixStack.scale(1.0F, 1.0F, 1.0F + f2 * 0.2F);
-                                pMatrixStack.mulPose(Vector3f.YN.rotationDegrees((float)k * 45.0F));
+                            if (item instanceof ArrowItem)
+                            {
+                                quaternion1 = Vector3f.ZP.rotationDegrees(-180.0F);
+                                quaternion.mul(Vector3f.XP.rotationDegrees((float)(-d3)));
+                            }
+                        }
+                        else if (vivecraftitemrendering == VivecraftItemRendering.Telescope)
+                        {
+                            quaternion1 = Vector3f.XP.rotationDegrees(0.0F);
+                            quaternion = Vector3f.XP.rotationDegrees(0.0F);
+                            d2 = 0.0D;
+                            d1 = 0.0D;
+                            d0 = 0.0D;
                         }
                     }
-                    else if (pPlayer.isAutoSpinAttack())
-                    {
-                        this.applyItemArmTransform(pMatrixStack, humanoidarm, pEquippedProgress);
-                        int i = flag1 ? 1 : -1;
-                        pMatrixStack.translate((double)((float)i * -0.4F), (double)0.8F, (double)0.3F);
-                        pMatrixStack.mulPose(Vector3f.YP.rotationDegrees((float)i * 65.0F));
-                        pMatrixStack.mulPose(Vector3f.ZP.rotationDegrees((float)i * -85.0F));
-                    }
-                    else
+                }
+                else
+                {
+                    quaternion = Vector3f.ZP.rotationDegrees(180.0F);
+                    quaternion.mul(Vector3f.XP.rotationDegrees(-135.0F));
+                    d8 = (double)0.4F;
+                    d0 += (double)0.08F;
+                    d2 += (double) - 0.08F;
+                }
+
+                if (pPlayer.swingingArm == pHand)
+                {
+                    this.transformFirstPersonVR(pMatrixStack, humanoidarm, pSwingProgress);
+                }
+
+                ItemRenderer itemrenderer = this.itemRenderer;
+                ItemRenderer.isfphand = true;
+                ItemTransforms.TransformType itemtransforms$transformtype = flag ? ItemTransforms.TransformType.FIRST_PERSON_RIGHT_HAND : (flag3 ? ItemTransforms.TransformType.FIRST_PERSON_LEFT_HAND : ItemTransforms.TransformType.FIRST_PERSON_RIGHT_HAND);
+
+                if (pStack.hasTag() && pStack.getTag().getInt("CustomModelData") > 0)
+                {
+                    itemtransforms$transformtype = ItemTransforms.TransformType.THIRD_PERSON_RIGHT_HAND;
+                    pMatrixStack.scale(0.7F, 0.7F, 0.7F);
+                }
+                else
+                {
+                    pMatrixStack.mulPose(quaternion1);
+                    pMatrixStack.translate(d0, d1, d2);
+                    pMatrixStack.mulPose(quaternion);
+                    pMatrixStack.scale((float)d8, (float)d8, (float)d8);
+                }
+
+                if (flag2)
+                {
+                    RenderSystem.disableCull();
+                    this.renderMap(pMatrixStack, pBuffer, pCombinedLight, pStack);
+                }
+                else if (vivecraftitemrendering == VivecraftItemRendering.Telescope)
+                {
+                    if (this.minecraft.currentPass != RenderPass.SCOPEL && this.minecraft.currentPass != RenderPass.SCOPER)
                     {
-                        float f6 = -0.4F * Mth.sin(Mth.sqrt(pSwingProgress) * (float)Math.PI);
-                        float f = 0.2F * Mth.sin(Mth.sqrt(pSwingProgress) * ((float)Math.PI * 2F));
-                        float f1 = -0.2F * Mth.sin(pSwingProgress * (float)Math.PI);
-                        int j = flag1 ? 1 : -1;
-                        pMatrixStack.translate((double)((float)j * f6), (double)f, (double)f1);
-                        this.applyItemArmTransform(pMatrixStack, humanoidarm, pEquippedProgress);
-                        this.applyItemArmAttackTransform(pMatrixStack, humanoidarm, pSwingProgress);
+                        pMatrixStack.pushPose();
+                        pMatrixStack.scale(0.625F, 0.625F, 0.625F);
+                        pMatrixStack.translate(flag ? -0.53D : -0.47D, -0.5D, -0.6D);
+                        //pMatrixStack.mulPose(Vector3f.XP.rotationDegrees(180.0F));
+                        this.minecraft.getBlockRenderer().getModelRenderer().renderModel(pMatrixStack.last(), pBuffer.getBuffer(Sheets.solidBlockSheet()), (BlockState)null, this.minecraft.getModelManager().getModel(TelescopeTracker.scopeModel), 0.5F, 0.5F, 1.0F, pCombinedLight, OverlayTexture.NO_OVERLAY);
+                        pMatrixStack.popPose();
                     }
 
-                    this.renderItem(pPlayer, pStack, flag1 ? ItemTransforms.TransformType.FIRST_PERSON_RIGHT_HAND : ItemTransforms.TransformType.FIRST_PERSON_LEFT_HAND, !flag1, pMatrixStack, pBuffer, pCombinedLight);
+                    pMatrixStack.pushPose();
+                    pMatrixStack.translate(flag ? -0.01875D : 0.01875D, 0.215D, -0.0626D);
+                    pMatrixStack.mulPose(Vector3f.XP.rotationDegrees(90.0F));
+                    pMatrixStack.mulPose(Vector3f.YP.rotationDegrees(180.0F));
+                    pMatrixStack.mulPose(Vector3f.ZP.rotationDegrees(180.0F));
+                    this.minecraft.gameRenderer.DrawScopeFB(pMatrixStack, pHand == InteractionHand.MAIN_HAND ? 0 : 1);
+                    pMatrixStack.popPose();
+                }
+                else
+                {
+                    this.renderItem(pPlayer, pStack, itemtransforms$transformtype, flag3, pMatrixStack, pBuffer, pCombinedLight);
                 }
 
+                itemrenderer = this.itemRenderer;
+                ItemRenderer.isfphand = false;
                 pMatrixStack.popPose();
             }
+
+            pMatrixStack.popPose();
         }
     }
 
@@ -679,4 +1067,41 @@
             return p_172932_ == InteractionHand.MAIN_HAND ? RENDER_MAIN_HAND_ONLY : RENDER_OFF_HAND_ONLY;
         }
     }
+
+    public float getEquipProgress(InteractionHand hand, float partialTicks)
+    {
+        return hand == InteractionHand.MAIN_HAND ? 1.0F - (this.oMainHandHeight + (this.mainHandHeight - this.oMainHandHeight) * partialTicks) : 1.0F - (this.oOffHandHeight + (this.offHandHeight - this.oOffHandHeight) * partialTicks);
+    }
+
+    public boolean isInsideOpaqueBlock(Vec3 in)
+    {
+        if (this.minecraft.level == null)
+        {
+            return false;
+        }
+        else
+        {
+            BlockPos blockpos = new BlockPos(in);
+            return this.minecraft.level.getBlockState(blockpos).isSolidRender(this.minecraft.level, blockpos);
+        }
+    }
+
+    @Nullable
+    public Triple<Float, BlockState, BlockPos> getNearOpaqueBlock(Vec3 in, double dist)
+    {
+        if (this.minecraft.level == null)
+        {
+            return null;
+        }
+        else
+        {
+            AABB aabb = new AABB(in.subtract(dist, dist, dist), in.add(dist, dist, dist));
+            Stream<BlockPos> stream = BlockPos.betweenClosedStream(aabb).filter((bp) ->
+            {
+                return this.minecraft.level.getBlockState(bp).isViewBlocking(this.minecraft.level, bp);
+            });
+            Optional<BlockPos> optional = stream.findFirst();
+            return optional.isPresent() ? Triple.of(1.0F, this.minecraft.level.getBlockState(optional.get()), optional.get()) : null;
+        }
+    }
 }
