--- a/net/minecraft/client/resources/DefaultClientPackResources.java
+++ b/net/minecraft/client/resources/DefaultClientPackResources.java
@@ -18,7 +18,8 @@
 
     public DefaultClientPackResources(PackMetadataSection pPackMetadata, AssetIndex pAssetIndex)
     {
-        super(pPackMetadata, "minecraft", "realms");
+    	//add
+        super(pPackMetadata, "minecraft", "realms", "vivecraft");
         this.assetIndex = pAssetIndex;
     }
 
