--- a/net/minecraft/client/renderer/entity/MobRenderer.java
+++ b/net/minecraft/client/renderer/entity/MobRenderer.java
@@ -3,6 +3,7 @@
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.blaze3d.vertex.VertexConsumer;
 import com.mojang.math.Matrix4f;
+import net.minecraft.client.Minecraft;
 import net.minecraft.client.model.EntityModel;
 import net.minecraft.client.renderer.LightTexture;
 import net.minecraft.client.renderer.MultiBufferSource;
@@ -61,6 +62,12 @@
         {
             pMatrixStack.pushPose();
             Vec3 vec3 = pLeashHolder.getRopeHoldPosition(pPartialTicks);
+
+            if (pLeashHolder == Minecraft.getInstance().player)
+            {
+                vec3 = Minecraft.getInstance().vrPlayer.vrdata_world_render.getController(0).getPosition();
+            }
+
             double d0 = (double)(Mth.lerp(pPartialTicks, pEntityLiving.yBodyRot, pEntityLiving.yBodyRotO) * ((float)Math.PI / 180F)) + (Math.PI / 2D);
             Vec3 vec31 = pEntityLiving.getLeashOffset();
             double d1 = Math.cos(d0) * vec31.z + Math.sin(d0) * vec31.x;
