--- a/net/minecraft/client/particle/ItemPickupParticle.java
+++ b/net/minecraft/client/particle/ItemPickupParticle.java
@@ -2,7 +2,9 @@
 
 import com.mojang.blaze3d.vertex.PoseStack;
 import com.mojang.blaze3d.vertex.VertexConsumer;
+
 import net.minecraft.client.Camera;
+import net.minecraft.client.Minecraft;
 import net.minecraft.client.multiplayer.ClientLevel;
 import net.minecraft.client.renderer.MultiBufferSource;
 import net.minecraft.client.renderer.RenderBuffers;
@@ -58,11 +60,20 @@
             Shaders.nextEntity(this.itemEntity);
         }
 
+
         float f = ((float)this.life + pPartialTicks) / 3.0F;
         f *= f;
-        double d0 = Mth.lerp((double)pPartialTicks, this.target.xOld, this.target.getX());
-        double d1 = Mth.lerp((double)pPartialTicks, this.target.yOld, this.target.getY()) + 0.5D;
-        double d2 = Mth.lerp((double)pPartialTicks, this.target.zOld, this.target.getZ());
+        Vec3 pos = this.target.position().add(0, 0.5f, 0);
+   
+        //Vivecraft
+        Minecraft mc = Minecraft.getInstance();
+        if (target== mc.player) {
+        	pos = mc.gameRenderer.getControllerRenderPos(0);
+        }   
+        //
+        double d0 = Mth.lerp((double)pPartialTicks, pos.x, pos.x);
+        double d1 = Mth.lerp((double)pPartialTicks, pos.y, pos.y);
+        double d2 = Mth.lerp((double)pPartialTicks, pos.z, pos.z);
         double d3 = Mth.lerp((double)f, this.itemEntity.getX(), d0);
         double d4 = Mth.lerp((double)f, this.itemEntity.getY(), d1);
         double d5 = Mth.lerp((double)f, this.itemEntity.getZ(), d2);
