--- a/com/mojang/blaze3d/platform/Window.java
+++ b/com/mojang/blaze3d/platform/Window.java
@@ -98,8 +98,10 @@
         GLFW.glfwWindowHint(139275, 221185);
         GLFW.glfwWindowHint(139266, 3);
         GLFW.glfwWindowHint(139267, 2);
-        GLFW.glfwWindowHint(139272, 204801);
-        GLFW.glfwWindowHint(139270, 1);
+        //Vivecraft keep compatibility profile?
+        GLFW.glfwWindowHint(139272, 204802);
+        //(was 204801)
+        //GLFW.glfwWindowHint(139270, 1);
         long i = 0L;
 
         if (Reflector.EarlyProgressVisualization_handOffWindow.exists())
@@ -154,7 +156,6 @@
         GL.createCapabilities();
         this.setMode();
         this.refreshFramebufferSize();
-        GLFW.glfwSetFramebufferSizeCallback(this.window, this::onFramebufferResize);
         GLFW.glfwSetWindowPosCallback(this.window, this::onMove);
         GLFW.glfwSetWindowSizeCallback(this.window, this::onResize);
         GLFW.glfwSetWindowFocusCallback(this.window, this::onFocus);
@@ -285,9 +286,9 @@
                 memorystack.close();
             }
         }
-        catch (IOException ioexception1)
+        catch (IOException ioexception)
         {
-            LOGGER.error("Couldn't set icon", (Throwable)ioexception1);
+            LOGGER.error("Couldn't set icon", (Throwable)ioexception);
         }
     }
 
@@ -364,7 +365,8 @@
     {
         RenderSystem.assertOnRenderThreadOrInit();
         this.vsync = pVsync;
-        GLFW.glfwSwapInterval(pVsync ? 1 : 0);
+        //Vivecraft no vsync
+        GLFW.glfwSwapInterval(0);
     }
 
     public void close()
@@ -387,15 +389,16 @@
     {
         if (p_85416_ == this.window)
         {
-            int i = this.getWidth();
-            int j = this.getHeight();
+        	//Vivecraft use screen sizes
+            int i = this.getScreenWidth();
+            int j = this.getScreenHeight();
 
             if (p_85417_ != 0 && p_85418_ != 0)
             {
                 this.framebufferWidth = p_85417_;
                 this.framebufferHeight = p_85418_;
 
-                if (this.getWidth() != i || this.getHeight() != j)
+                if (this.getScreenWidth() != i || this.getScreenHeight() != j)
                 {
                     this.eventHandler.resizeDisplay();
                 }
@@ -423,11 +426,15 @@
             });
         }
     }
-
-    private void onResize(long p_85428_, int p_85429_, int p_85430_)
+	//vivecraft dont resize on minimize
+    private void onResize(long pWindowPointer, int p_85429_, int pWindowWidth)
     {
-        this.width = p_85429_;
-        this.height = p_85430_;
+        if (pWindowWidth * p_85429_ != 0)
+        {
+            this.width = p_85429_;
+            this.height = pWindowWidth;
+            this.eventHandler.resizeDisplay();
+        }
     }
 
     private void onFocus(long p_85393_, boolean p_85394_)
@@ -575,13 +582,12 @@
     {
         int i;
 
-        for (i = 1; i != pGuiScale && i < this.framebufferWidth && i < this.framebufferHeight && this.framebufferWidth / (i + 1) >= 320 && this.framebufferHeight / (i + 1) >= 240; ++i)
+        for (i = 1; i != pGuiScale && i < this.width && i < this.height && this.width / (i + 1) >= 320 && this.height / (i + 1) >= 240; ++i)
         {
         }
 
         if (pForceUnicode && i % 2 != 0)
         {
-            ++i;
         }
 
         return i;
@@ -590,10 +596,10 @@
     public void setGuiScale(double pScaleFactor)
     {
         this.guiScale = pScaleFactor;
-        int i = (int)((double)this.framebufferWidth / pScaleFactor);
-        this.guiScaledWidth = (double)this.framebufferWidth / pScaleFactor > (double)i ? i + 1 : i;
-        int j = (int)((double)this.framebufferHeight / pScaleFactor);
-        this.guiScaledHeight = (double)this.framebufferHeight / pScaleFactor > (double)j ? j + 1 : j;
+        int i = (int)((double)this.width / pScaleFactor);
+        this.guiScaledWidth = (double)this.width / pScaleFactor > (double)i ? i + 1 : i;
+        int j = (int)((double)this.height / pScaleFactor);
+        this.guiScaledHeight = (double)this.height / pScaleFactor > (double)j ? j + 1 : j;
     }
 
     public void setTitle(String pTitle)
@@ -613,12 +619,12 @@
 
     public int getWidth()
     {
-        return this.framebufferWidth;
+        return Minecraft.getInstance().getMainRenderTarget().viewWidth;
     }
 
     public int getHeight()
     {
-        return this.framebufferHeight;
+        return Minecraft.getInstance().getMainRenderTarget().viewHeight;
     }
 
     public void setWidth(int pFramebufferWidth)
@@ -677,9 +683,9 @@
         InputConstants.updateRawMouseInput(this.window, pEnableRawMouseMotion);
     }
 
-    public void resizeFramebuffer(int width, int height)
+    public void resizeFramebuffer(int p_resizeFramebuffer_1_, int p_resizeFramebuffer_2_)
     {
-        this.onFramebufferResize(this.window, width, height);
+        this.onFramebufferResize(this.window, p_resizeFramebuffer_1_, p_resizeFramebuffer_2_);
     }
 
     public boolean isClosed()
