--- a/com/mojang/blaze3d/pipeline/MainTarget.java
+++ b/com/mojang/blaze3d/pipeline/MainTarget.java
@@ -10,9 +10,9 @@
 
 public class MainTarget extends RenderTarget
 {
-    public static final int DEFAULT_WIDTH = 854;
-    public static final int DEFAULT_HEIGHT = 480;
-    static final MainTarget.Dimension DEFAULT_DIMENSIONS = new MainTarget.Dimension(854, 480);
+    public static final int DEFAULT_WIDTH = 1280;
+    public static final int DEFAULT_HEIGHT = 720;
+    static final MainTarget.Dimension DEFAULT_DIMENSIONS = new MainTarget.Dimension(DEFAULT_WIDTH, DEFAULT_HEIGHT);
 
     public MainTarget(int pWidth, int pHeight)
     {
