--- a/com/mojang/blaze3d/pipeline/TextureTarget.java
+++ b/com/mojang/blaze3d/pipeline/TextureTarget.java
@@ -10,4 +10,11 @@
         RenderSystem.assertOnRenderThreadOrInit();
         this.resize(pWidth, pHeight, pClearError);
     }
+    
+    //Vivecraft extended constructor. TODO: custom rendertarget impl?
+    public TextureTarget(String name, int width, int height, boolean usedepth, boolean onMac, int texid, boolean depthtex, boolean linearFilter) {
+    	super(name, width, height, usedepth, onMac,texid, depthtex, linearFilter);
+    	this.setClearColor(0.0F, 0.0F, 0.0F, 0.0F);
+    }
+
 }
