--- a/net/minecraft/client/model/geom/ModelPart.java
+++ b/net/minecraft/client/model/geom/ModelPart.java
@@ -439,7 +439,7 @@
 
     public static class Cube
     {
-        private final ModelPart.Polygon[] polygons;
+        public final ModelPart.Polygon[] polygons;
         public final float minX;
         public final float minY;
         public final float minZ;
@@ -606,10 +606,10 @@
         }
     }
 
-    static class Polygon
+    public static class Polygon
     {
-        public final ModelPart.Vertex[] vertices;
-        public final Vector3f normal;
+    	public final ModelPart.Vertex[] vertices;
+    	public final Vector3f normal;
 
         public Polygon(ModelPart.Vertex[] p_104362_, float p_104363_, float p_104364_, float p_104365_, float p_104366_, float p_104367_, float p_104368_, boolean p_104369_, Direction p_104370_)
         {
@@ -659,7 +659,7 @@
         }
     }
 
-    static class Vertex
+    public static class Vertex
     {
         public final Vector3f pos;
         public final float u;
