--- a/net/minecraft/client/tutorial/OpenInventoryTutorialStep.java
+++ b/net/minecraft/client/tutorial/OpenInventoryTutorialStep.java
@@ -26,13 +26,10 @@
         {
             this.tutorial.setStep(TutorialSteps.NONE);
         }
-        else
+        else if (this.timeWaiting >= 600 && this.toast == null)
         {
-            if (this.timeWaiting >= 600 && this.toast == null)
-            {
-                this.toast = new TutorialToast(TutorialToast.Icons.RECIPE_BOOK, TITLE, DESCRIPTION, false);
-                this.tutorial.getMinecraft().getToasts().addToast(this.toast);
-            }
+            this.toast = new TutorialToast(TutorialToast.Icons.RECIPE_BOOK, TITLE, DESCRIPTION, false);
+            this.tutorial.getInstance().getToasts().addToast(this.toast);
         }
     }
 
