--- a/net/minecraft/client/multiplayer/ClientPacketListener.java
+++ b/net/minecraft/client/multiplayer/ClientPacketListener.java
@@ -250,9 +250,21 @@
 import net.minecraft.world.level.GameType;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LightLayer;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.entity.BannerBlockEntity;
+import net.minecraft.world.level.block.entity.BeaconBlockEntity;
+import net.minecraft.world.level.block.entity.BedBlockEntity;
+import net.minecraft.world.level.block.entity.BeehiveBlockEntity;
 import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.block.entity.CampfireBlockEntity;
 import net.minecraft.world.level.block.entity.CommandBlockEntity;
+import net.minecraft.world.level.block.entity.ConduitBlockEntity;
+import net.minecraft.world.level.block.entity.JigsawBlockEntity;
 import net.minecraft.world.level.block.entity.SignBlockEntity;
+import net.minecraft.world.level.block.entity.SkullBlockEntity;
+import net.minecraft.world.level.block.entity.SpawnerBlockEntity;
+import net.minecraft.world.level.block.entity.StructureBlockEntity;
+import net.minecraft.world.level.block.entity.TheEndGatewayBlockEntity;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.border.WorldBorder;
 import net.minecraft.world.level.chunk.DataLayer;
@@ -273,8 +285,14 @@
 import net.minecraft.world.scores.Scoreboard;
 import net.minecraft.world.scores.Team;
 import net.minecraft.world.scores.criteria.ObjectiveCriteria;
+import net.optifine.reflect.Reflector;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.vivecraft.api.NetworkHelper;
+import org.vivecraft.provider.ControllerType;
+import org.vivecraft.provider.openvr_jna.control.VivecraftMovementInput;
+import org.vivecraft.render.PlayerModelController;
+import org.vivecraft.settings.VRSettings;
 
 public class ClientPacketListener implements ClientGamePacketListener
 {
@@ -311,6 +329,8 @@
         this.advancements = new ClientAdvancements(p_194193_);
         this.suggestionsProvider = new ClientSuggestionProvider(this, p_194193_);
         this.telemetryManager = p_194197_;
+        NetworkHelper.resetServerSettings();
+        NetworkHelper.displayedChatMessage = false;
     }
 
     public ClientSuggestionProvider getSuggestionsProvider()
@@ -366,10 +386,15 @@
 
         this.minecraft.debugRenderer.clear();
         this.minecraft.player.resetPos();
+        //Forge
+        Reflector.call(Reflector.ClientHooks_firePlayerLogin, this.minecraft.gameMode, this.minecraft.player, this.minecraft.getConnection().getConnection());
+        //
         int i = pPacket.playerId();
         this.minecraft.player.setId(i);
         this.level.addPlayer(i, this.minecraft.player);
-        this.minecraft.player.input = new KeyboardInput(this.minecraft.options);
+        //Vivecraft
+        this.minecraft.player.input = new VivecraftMovementInput(this.minecraft.options);
+        //
         this.minecraft.gameMode.adjustPlayer(this.minecraft.player);
         this.minecraft.cameraEntity = this.minecraft.player;
         this.minecraft.setScreen(new ReceivingLevelScreen());
@@ -377,10 +402,14 @@
         this.minecraft.player.setShowDeathScreen(pPacket.showDeathScreen());
         this.minecraft.gameMode.setLocalMode(pPacket.gameType(), pPacket.previousGameType());
         this.minecraft.options.setServerRenderDistance(pPacket.chunkRadius());
+        //Forge
+        Reflector.call(Reflector.NetworkHooks_sendMCRegistryPackets, connection, "PLAY_TO_SERVER");
+        //
         this.minecraft.options.broadcastOptions();
         this.connection.send(new ServerboundCustomPayloadPacket(ServerboundCustomPayloadPacket.BRAND, (new FriendlyByteBuf(Unpooled.buffer())).writeUtf(ClientBrandRetriever.getClientModName())));
         this.minecraft.getGame().onStartGameSession();
-        this.telemetryManager.onPlayerInfoReceived(pPacket.gameType(), pPacket.hardcore());
+        NetworkHelper.vivePlayers.clear();
+        NetworkHelper.sendVersionInfo();
     }
 
     public void handleAddEntity(ClientboundAddEntityPacket pPacket)
@@ -649,9 +678,9 @@
     {
         PacketUtils.ensureRunningOnSameThread(pPacket, this, this.minecraft);
         int i = 19 | (pPacket.shouldSuppressLightUpdates() ? 128 : 0);
-        pPacket.runUpdates((p_104922_, p_104923_) ->
+        pPacket.runUpdates((p_243492_2_, p_243492_3_) ->
         {
-            this.level.setBlock(p_104922_, p_104923_, i);
+            this.level.setBlock(p_243492_2_, p_243492_3_, i);
         });
     }
 
@@ -735,6 +764,9 @@
     public void handleDisconnect(ClientboundDisconnectPacket pPacket)
     {
         this.connection.disconnect(pPacket.getReason());
+        this.minecraft.vrPlayer.setTeleportSupported(false);
+        this.minecraft.vrPlayer.setTeleportOverride(false);
+        this.minecraft.vrSettings.overrides.resetAll();
     }
 
     public void onDisconnect(Component pReason)
@@ -809,7 +841,39 @@
     public void handleChat(ClientboundChatPacket pPacket)
     {
         PacketUtils.ensureRunningOnSameThread(pPacket, this, this.minecraft);
-        this.minecraft.gui.handleChat(pPacket.getType(), pPacket.getMessage(), pPacket.getSender());
+        //Forge
+        Component component = (Component)Reflector.call(Reflector.ForgeEventFactory_onClientChat, pPacket.getType(), pPacket.getMessage(), pPacket.getSender());
+
+        if (component == null)
+        {
+            this.minecraft.gui.handleChat(pPacket.getType(), pPacket.getMessage(), pPacket.getSender());
+        }
+        else
+        {
+            this.minecraft.gui.handleChat(pPacket.getType(), component, pPacket.getSender());
+        }
+        //
+        
+        Minecraft minecraft = Minecraft.getInstance();
+        String s = minecraft.player.lastMsg;
+        minecraft.player.lastMsg = null;
+
+        if (minecraft.player == null || s == null || !pPacket.getMessage().getString().contains(s))
+        {
+            if (minecraft.vrSettings.chatNotifications != VRSettings.ChatNotifications.NONE)
+            {
+                if ((minecraft.vrSettings.chatNotifications == VRSettings.ChatNotifications.HAPTIC || minecraft.vrSettings.chatNotifications == VRSettings.ChatNotifications.BOTH) && !minecraft.vrSettings.seated)
+                {
+                    minecraft.vr.triggerHapticPulse(ControllerType.LEFT, 0.2F, 1000.0F, 1.0F);
+                }
+
+                if (minecraft.vrSettings.chatNotifications == VRSettings.ChatNotifications.SOUND || minecraft.vrSettings.chatNotifications == VRSettings.ChatNotifications.BOTH)
+                {
+                    Vec3 vec3 = minecraft.vrPlayer.vrdata_world_pre.getController(1).getPosition();
+                    minecraft.level.playLocalSound(vec3.x(), vec3.y(), vec3.z(), Registry.SOUND_EVENT.get(new ResourceLocation(minecraft.vrSettings.chatNotificationSound)), SoundSource.NEUTRAL, 0.3F, 0.1F, false);
+                }
+            }
+        }
     }
 
     public void handleAnimate(ClientboundAnimatePacket pPacket)
@@ -919,7 +983,8 @@
 
                     if (entity1 == this.minecraft.player && !flag)
                     {
-                        this.minecraft.gui.setOverlayMessage(new TranslatableComponent("mount.onboard", this.minecraft.options.keyShift.getTranslatedKeyMessage()), false);
+                        this.minecraft.gui.setOverlayMessage(new TranslatableComponent("mount.onboard", this.minecraft.options.keyShift.getName()), false);
+                        this.minecraft.options.keyShift.getTranslatedKeyMessage();
                     }
                 }
             }
@@ -1017,6 +1082,8 @@
             this.level.addMapData(map);
             this.minecraft.setLevel(this.level);
             this.minecraft.setScreen(new ReceivingLevelScreen());
+            NetworkHelper.resetServerSettings();
+            NetworkHelper.sendVersionInfo();
         }
 
         String s = localplayer.getServerBrand();
@@ -1037,12 +1104,15 @@
         {
             localplayer1.getAttributes().assignValues(localplayer.getAttributes());
         }
-
+        //Forge
+        localplayer1.updateSyncFields(localplayer); // Forge: fix MC-10657
+        //
         localplayer1.resetPos();
         localplayer1.setServerBrand(s);
+        Reflector.call(Reflector.ClientHooks_firePlayerRespawn, this.minecraft.gameMode, localplayer, localplayer1, localplayer1.connection.getConnection());
         this.level.addPlayer(i, localplayer1);
         localplayer1.setYRot(-180.0F);
-        localplayer1.input = new KeyboardInput(this.minecraft.options);
+        localplayer1.input = new VivecraftMovementInput(this.minecraft.options);
         this.minecraft.gameMode.adjustPlayer(localplayer1);
         localplayer1.setReducedDebugInfo(localplayer.isReducedDebugInfo());
         localplayer1.setShowDeathScreen(localplayer.shouldShowDeathScreen());
@@ -1168,22 +1238,27 @@
 
     public void handleBlockEntityData(ClientboundBlockEntityDataPacket pPacket)
     {
-        PacketUtils.ensureRunningOnSameThread(pPacket, this, this.minecraft);
-        BlockPos blockpos = pPacket.getPos();
-        this.minecraft.level.getBlockEntity(blockpos, pPacket.getType()).ifPresent((p_194239_) ->
-        {
-            CompoundTag compoundtag = pPacket.getTag();
-
-            if (compoundtag != null)
-            {
-                p_194239_.load(compoundtag);
-            }
-
-            if (p_194239_ instanceof CommandBlockEntity && this.minecraft.screen instanceof CommandBlockEditScreen)
-            {
-                ((CommandBlockEditScreen)this.minecraft.screen).updateGui();
-            }
-        });
+    	PacketUtils.ensureRunningOnSameThread(pPacket, this, this.minecraft);
+    	BlockPos blockpos = pPacket.getPos();
+    	this.minecraft.level.getBlockEntity(blockpos, pPacket.getType()).ifPresent((p_194239_) ->
+    	{
+    		//Forge
+    		if(Reflector.IForgeTileEntity_handleUpdateTag.exists()) {
+    			Reflector.call(p_194239_, Reflector.IForgeTileEntity_onDataPacket, connection, pPacket);
+    		} else {
+    		//
+    			CompoundTag compoundtag = pPacket.getTag();
+
+    			if (compoundtag != null)
+    			{
+    				p_194239_.load(compoundtag);
+    			}
+    		}
+    		if (p_194239_ instanceof CommandBlockEntity && this.minecraft.screen instanceof CommandBlockEditScreen)
+    		{
+    			((CommandBlockEditScreen)this.minecraft.screen).updateGui();
+    		}
+    	});
     }
 
     public void handleContainerSetData(ClientboundContainerSetDataPacket pPacket)
@@ -1408,6 +1483,9 @@
         clientrecipebook.setupCollections(this.recipeManager.getRecipes());
         clientrecipebook.getCollections().forEach(mutablesearchtree::add);
         mutablesearchtree.refresh();
+        //Forge
+        Reflector.call(Reflector.ForgeHooksClient_onRecipesUpdated, this.recipeManager);
+        //
     }
 
     public void handleLookAt(ClientboundPlayerLookAtPacket pPacket)
@@ -1508,7 +1586,7 @@
 
         if (entity instanceof LivingEntity)
         {
-            MobEffect mobeffect = MobEffect.byId(pPacket.getEffectId());
+            MobEffect mobeffect = MobEffect.byId(pPacket.getEffectId() & 255);
 
             if (mobeffect != null)
             {
@@ -1532,6 +1610,11 @@
         }
         else
         {
+        	//Forge
+        	if(Reflector.ITagCollectionSupplier_reinjectOptionalTags.exists()) {
+        		tagcontainer = (TagContainer) Reflector.call(Reflector.ITagCollectionSupplier_reinjectOptionalTags, tagcontainer);
+        	}
+        	//
             this.tags = tagcontainer;
 
             if (!this.connection.isMemoryConnection())
@@ -1886,16 +1969,16 @@
             {
                 throw new URISyntaxException(pUrl, "Wrong protocol");
             }
-            else if (!flag || !pUrl.contains("..") && pUrl.endsWith("/resources.zip"))
+            else if (flag && (pUrl.contains("..") || !pUrl.endsWith("/resources.zip")))
             {
-                return true;
+                throw new URISyntaxException(pUrl, "Invalid levelstorage resourcepack path");
             }
             else
             {
-                throw new URISyntaxException(pUrl, "Invalid levelstorage resourcepack path");
+                return true;
             }
         }
-        catch (URISyntaxException urisyntaxexception)
+        catch (URISyntaxException urisyntaxexception1)
         {
             this.send(ServerboundResourcePackPacket.Action.FAILED_DOWNLOAD);
             return false;
@@ -1967,6 +2050,7 @@
         PacketUtils.ensureRunningOnSameThread(pPacket, this, this.minecraft);
         ResourceLocation resourcelocation = pPacket.getIdentifier();
         FriendlyByteBuf friendlybytebuf = null;
+        boolean release = true;
 
         try
         {
@@ -2243,14 +2327,170 @@
                 int i5 = friendlybytebuf.readVarInt();
                 this.minecraft.debugRenderer.gameEventListenerRenderer.trackListener(positionsource, i5);
             }
+            else if (resourcelocation.getNamespace().equalsIgnoreCase("vivecraft"))
+            {
+                if (resourcelocation.getPath().equalsIgnoreCase("data"))
+                {
+                    byte b0 = friendlybytebuf.readByte();
+                    NetworkHelper.PacketDiscriminators networkhelper$packetdiscriminators = NetworkHelper.PacketDiscriminators.values()[b0];
+
+                    switch (networkhelper$packetdiscriminators)
+                    {
+                        case VERSION:
+                            String s11 = friendlybytebuf.readUtf(1024);
+                            this.minecraft.vrPlayer.setTeleportSupported(true);
+                            this.minecraft.vrPlayer.teleportWarningTimer = -1;
+
+                            if (!NetworkHelper.displayedChatMessage)
+                            {
+                                NetworkHelper.displayedChatMessage = true;
+                                this.minecraft.gui.getChat().addMessage(new TranslatableComponent("vivecraft.messages.serverplugin", s11));
+                            }
+
+                            if (Minecraft.getInstance().vrSettings.manualCalibration == -1.0F && !Minecraft.getInstance().vrSettings.seated)
+                            {
+                                this.minecraft.gui.getChat().addMessage(new TranslatableComponent("vivecraft.messages.calibrateheight"));
+                            }
+
+                            break;
+
+                        case REQUESTDATA:
+                            NetworkHelper.serverWantsData = true;
+                            break;
+
+                        case CLIMBING:
+                            NetworkHelper.serverAllowsClimbey = friendlybytebuf.readBoolean();
+
+                            if (friendlybytebuf.readableBytes() > 0)
+                            {
+                                Minecraft.getInstance().climbTracker.serverblockmode = friendlybytebuf.readByte();
+                                Minecraft.getInstance().climbTracker.blocklist.clear();
+
+                                while (friendlybytebuf.readableBytes() > 0)
+                                {
+                                    String s12 = friendlybytebuf.readUtf(16384);
+                                    Block block = Registry.BLOCK.get(new ResourceLocation(s12));
+
+                                    if (block != null)
+                                    {
+                                        Minecraft.getInstance().climbTracker.blocklist.add(block);
+                                    }
+                                }
+                            }
+
+                            break;
+
+                        case TELEPORT:
+                            NetworkHelper.serverSupportsDirectTeleport = true;
+                            break;
+
+                        case UBERPACKET:
+                            Long olong = friendlybytebuf.readLong();
+                            Long olong1 = friendlybytebuf.readLong();
+                            byte[] abyte = new byte[29];
+                            byte[] abyte1 = new byte[29];
+                            byte[] abyte2 = new byte[29];
+                            friendlybytebuf.readBytes(29).getBytes(0, abyte);
+                            friendlybytebuf.readBytes(29).getBytes(0, abyte1);
+                            friendlybytebuf.readBytes(29).getBytes(0, abyte2);
+                            UUID uuid2 = new UUID(olong, olong1);
+                            float f3 = 1.0F;
+                            float f4 = 1.0F;
+
+                            if (friendlybytebuf.isReadable())
+                            {
+                                f3 = friendlybytebuf.readFloat();
+                            }
+
+                            if (friendlybytebuf.isReadable())
+                            {
+                                f4 = friendlybytebuf.readFloat();
+                            }
+
+                            PlayerModelController.getInstance().Update(uuid2, abyte, abyte1, abyte2, f3, f4);
+                            break;
+
+                        case SETTING_OVERRIDE:
+                            while (friendlybytebuf.readableBytes() > 0)
+                            {
+                                String s13 = friendlybytebuf.readUtf(16384);
+                                String s14 = friendlybytebuf.readUtf(16384);
+                                String[] astring = s13.split("\\.", 2);
+
+                                if (this.minecraft.vrSettings.overrides.hasSetting(astring[0]))
+                                {
+                                    VRSettings.ServerOverrides.Setting vrsettings$serveroverrides$setting = this.minecraft.vrSettings.overrides.getSetting(astring[0]);
+
+                                    try
+                                    {
+                                        if (astring.length > 1)
+                                        {
+                                            String s15 = astring[1];
+
+                                            switch (s15)
+                                            {
+                                                case "min":
+                                                    vrsettings$serveroverrides$setting.setValueMin(Float.parseFloat(s14));
+                                                    break;
+
+                                                case "max":
+                                                    vrsettings$serveroverrides$setting.setValueMax(Float.parseFloat(s14));
+                                            }
+                                        }
+                                        else
+                                        {
+                                            Object object = vrsettings$serveroverrides$setting.getOriginalValue();
+
+                                            if (object instanceof Boolean)
+                                            {
+                                                vrsettings$serveroverrides$setting.setValue(s14.equals("true"));
+                                            }
+                                            else if (!(object instanceof Integer) && !(object instanceof Byte) && !(object instanceof Short))
+                                            {
+                                                if (!(object instanceof Float) && !(object instanceof Double))
+                                                {
+                                                    vrsettings$serveroverrides$setting.setValue(s14);
+                                                }
+                                                else
+                                                {
+                                                    vrsettings$serveroverrides$setting.setValue(Float.parseFloat(s14));
+                                                }
+                                            }
+                                            else
+                                            {
+                                                vrsettings$serveroverrides$setting.setValue(Integer.parseInt(s14));
+                                            }
+                                        }
+
+                                        System.out.println("Server setting override: " + s13 + " = " + s14);
+                                    }
+                                    catch (Exception exception)
+                                    {
+                                        exception.printStackTrace();
+                                    }
+                                }
+                            }
+
+                            break;
+
+                        case CRAWL:
+                            NetworkHelper.serverAllowsCrawling = true;
+                    }
+                }
+            }
             else
             {
-                LOGGER.warn("Unknown custom packed identifier: {}", (Object)resourcelocation);
+            	release = false;
+
+                if (!Reflector.callBoolean(Reflector.NetworkHooks_onCustomPayload, pPacket, this.connection))
+                {
+                    LOGGER.warn("Unknown custom packet identifier: {}", (Object)resourcelocation);
+                }
             }
         }
         finally
         {
-            if (friendlybytebuf != null)
+            if (friendlybytebuf != null && release)
             {
                 friendlybytebuf.release();
             }
@@ -2439,27 +2679,25 @@
             {
                 throw new IllegalStateException("Server tried to update attributes of a non-living entity (actually: " + entity + ")");
             }
-            else
+
+            AttributeMap attributemap = ((LivingEntity)entity).getAttributes();
+
+            for (ClientboundUpdateAttributesPacket.AttributeSnapshot clientboundupdateattributespacket$attributesnapshot : pPacket.getValues())
             {
-                AttributeMap attributemap = ((LivingEntity)entity).getAttributes();
+                AttributeInstance attributeinstance = attributemap.getInstance(clientboundupdateattributespacket$attributesnapshot.getAttribute());
 
-                for (ClientboundUpdateAttributesPacket.AttributeSnapshot clientboundupdateattributespacket$attributesnapshot : pPacket.getValues())
+                if (attributeinstance == null)
+                {
+                    LOGGER.warn("Entity {} does not have attribute {}", entity, Registry.ATTRIBUTE.getKey(clientboundupdateattributespacket$attributesnapshot.getAttribute()));
+                }
+                else
                 {
-                    AttributeInstance attributeinstance = attributemap.getInstance(clientboundupdateattributespacket$attributesnapshot.getAttribute());
+                    attributeinstance.setBaseValue(clientboundupdateattributespacket$attributesnapshot.getBase());
+                    attributeinstance.removeModifiers();
 
-                    if (attributeinstance == null)
+                    for (AttributeModifier attributemodifier : clientboundupdateattributespacket$attributesnapshot.getModifiers())
                     {
-                        LOGGER.warn("Entity {} does not have attribute {}", entity, Registry.ATTRIBUTE.getKey(clientboundupdateattributespacket$attributesnapshot.getAttribute()));
-                    }
-                    else
-                    {
-                        attributeinstance.setBaseValue(clientboundupdateattributespacket$attributesnapshot.getBase());
-                        attributeinstance.removeModifiers();
-
-                        for (AttributeModifier attributemodifier : clientboundupdateattributespacket$attributesnapshot.getModifiers())
-                        {
-                            attributeinstance.addTransientModifier(attributemodifier);
-                        }
+                        attributeinstance.addTransientModifier(attributemodifier);
                     }
                 }
             }
@@ -2648,4 +2886,17 @@
     {
         return this.registryAccess;
     }
+
+    boolean tryParseInt(String value)
+    {
+        try
+        {
+            Integer.parseInt(value);
+            return true;
+        }
+        catch (NumberFormatException numberformatexception)
+        {
+            return false;
+        }
+    }
 }
