--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -3,6 +3,10 @@
 import com.google.common.base.MoreObjects;
 import java.util.UUID;
 import javax.annotation.Nullable;
+
+import org.vivecraft.api.NetworkHelper;
+import org.vivecraft.api.ServerVivePlayer;
+
 import net.minecraft.core.BlockPos;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.protocol.Packet;
@@ -120,9 +124,9 @@
 
         if (entity != null)
         {
-            for (Entity entity1 : this.level.getEntities(this, this.getBoundingBox().expandTowards(this.getDeltaMovement()).inflate(1.0D), (p_37272_) ->
+            for (Entity entity1 : this.level.getEntities(this, this.getBoundingBox().expandTowards(this.getDeltaMovement()).inflate(1.0D), (p_234613_0_) ->
         {
-            return !p_37272_.isSpectator() && p_37272_.isPickable();
+            return !p_234613_0_.isSpectator() && p_234613_0_.isPickable();
             }))
             {
                 if (entity1.getRootVehicle() == entity.getRootVehicle())
@@ -148,12 +152,29 @@
 
     public void shootFromRotation(Entity pProjectile, float pX, float pY, float pZ, float pVelocity, float pInaccuracy)
     {
-        float f = -Mth.sin(pY * ((float)Math.PI / 180F)) * Mth.cos(pX * ((float)Math.PI / 180F));
-        float f1 = -Mth.sin((pX + pZ) * ((float)Math.PI / 180F));
-        float f2 = Mth.cos(pY * ((float)Math.PI / 180F)) * Mth.cos(pX * ((float)Math.PI / 180F));
-        this.shoot((double)f, (double)f1, (double)f2, pVelocity, pInaccuracy);
-        Vec3 vec3 = pProjectile.getDeltaMovement();
-        this.setDeltaMovement(this.getDeltaMovement().add(vec3.x, pProjectile.isOnGround() ? 0.0D : vec3.y, vec3.z));
+        ServerVivePlayer serverviveplayer = NetworkHelper.vivePlayers.get(pProjectile.getUUID());
+
+        if (serverviveplayer != null && serverviveplayer.isVR())
+        {
+            Vec3 vec3 = serverviveplayer.getControllerDir(serverviveplayer.activeHand);
+
+            if (this instanceof AbstractArrow && !(this instanceof ThrownTrident) && !serverviveplayer.isSeated() && serverviveplayer.getDraw() > 0.0F)
+            {
+                vec3 = serverviveplayer.getControllerPos(1, (Player)pProjectile).subtract(serverviveplayer.getControllerPos(0, (Player)pProjectile)).normalize();
+                pVelocity *= serverviveplayer.getDraw();
+                ((AbstractArrow)this).setBaseDamage(((AbstractArrow)this).getBaseDamage() * 2.0D);
+            }
+
+            pX = -((float)Math.toDegrees(Math.asin(vec3.y / vec3.length())));
+            pY = (float)Math.toDegrees(Math.atan2(-vec3.x, vec3.z));
+        }
+
+        float f2 = -Mth.sin(pY * ((float)Math.PI / 180F)) * Mth.cos(pX * ((float)Math.PI / 180F));
+        float f = -Mth.sin((pX + pZ) * ((float)Math.PI / 180F));
+        float f1 = Mth.cos(pY * ((float)Math.PI / 180F)) * Mth.cos(pX * ((float)Math.PI / 180F));
+        this.shoot((double)f2, (double)f, (double)f1, pVelocity, pInaccuracy);
+        Vec3 vec31 = pProjectile.getDeltaMovement();
+        this.setDeltaMovement(this.getDeltaMovement().add(vec31.x, pProjectile.isOnGround() ? 0.0D : vec31.y, vec31.z));
     }
 
     protected void onHit(HitResult pResult)
