--- a/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -41,6 +41,9 @@
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import net.optifine.reflect.Reflector;
+import org.vivecraft.api.NetworkHelper;
+import org.vivecraft.api.ServerVivePlayer;
 
 public abstract class AbstractArrow extends Projectile
 {
@@ -83,6 +86,22 @@
 
         if (p_36718_ instanceof Player)
         {
+            ServerVivePlayer serverviveplayer = NetworkHelper.vivePlayers.get(p_36718_.getUUID());
+
+            if (serverviveplayer != null && serverviveplayer.isVR())
+            {
+                Vec3 vec3 = serverviveplayer.getControllerPos(serverviveplayer.activeHand, (Player)p_36718_);
+                Vec3 vec31 = serverviveplayer.getControllerDir(serverviveplayer.activeHand);
+
+                if (!serverviveplayer.isSeated() && serverviveplayer.getDraw() > 0.0F)
+                {
+                    vec31 = serverviveplayer.getControllerPos(1, (Player)p_36718_).subtract(serverviveplayer.getControllerPos(0, (Player)p_36718_)).normalize();
+                    vec3 = serverviveplayer.getControllerPos(0, (Player)p_36718_);
+                }
+
+                this.setPos(vec3.x + vec31.x, vec3.y + vec31.y, vec3.z + vec31.z);
+            }
+
             this.pickup = AbstractArrow.Pickup.ALLOWED;
         }
     }
@@ -146,8 +165,18 @@
 
         BlockPos blockpos = this.blockPosition();
         BlockState blockstate = this.level.getBlockState(blockpos);
+        boolean flag1;
+
+        if (Reflector.IForgeBlockState_isAir2.exists())
+        {
+            flag1 = Reflector.callBoolean(blockstate, Reflector.IForgeBlockState_isAir2, this.level, blockpos);
+        }
+        else
+        {
+            flag1 = blockstate.isAir();
+        }
 
-        if (!blockstate.isAir() && !flag)
+        if (!flag1 && !flag)
         {
             VoxelShape voxelshape = blockstate.getCollisionShape(this.level, blockpos);
 
@@ -222,7 +251,14 @@
                     }
                 }
 
-                if (hitresult != null && !flag)
+                boolean flag2 = true;
+
+                if (hitresult != null && Reflector.ForgeEventFactory_onProjectileImpactA.exists())
+                {
+                    flag2 = hitresult.getType() != HitResult.Type.MISS && !Reflector.callBoolean(Reflector.ForgeEventFactory_onProjectileImpactA, this, hitresult);
+                }
+
+                if (hitresult != null && !flag && flag2)
                 {
                     this.onHit(hitresult);
                     this.hasImpulse = true;
