--- a/com/mojang/math/Matrix4f.java
+++ b/com/mojang/math/Matrix4f.java
@@ -2,26 +2,27 @@
 
 import java.nio.FloatBuffer;
 import java.util.Random;
+import net.minecraft.world.phys.Vec3;
 
 public final class Matrix4f
 {
     private static final int ORDER = 4;
-    protected float m00;
-    protected float m01;
-    protected float m02;
-    protected float m03;
-    protected float m10;
-    protected float m11;
-    protected float m12;
-    protected float m13;
-    protected float m20;
-    protected float m21;
-    protected float m22;
-    protected float m23;
-    protected float m30;
-    protected float m31;
-    protected float m32;
-    protected float m33;
+    public float m00;
+    public float m01;
+    public float m02;
+    public float m03;
+    public float m10;
+    public float m11;
+    public float m12;
+    public float m13;
+    public float m20;
+    public float m21;
+    public float m22;
+    public float m23;
+    public float m30;
+    public float m31;
+    public float m32;
+    public float m33;
 
     public Matrix4f()
     {
@@ -808,4 +809,10 @@
         this.m13 = y;
         this.m23 = z;
     }
+    //Vivecraft
+    public Vec3 translationComponent()
+    {
+    	return new Vec3((double)this.m30, (double)this.m31, (double)this.m32);
+    }
+    //
 }
