--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -5,6 +5,7 @@
 import com.mojang.datafixers.util.Either;
 import java.util.Collection;
 import java.util.List;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.OptionalInt;
 import java.util.Random;
@@ -17,11 +18,14 @@
 import net.minecraft.ReportedException;
 import net.minecraft.Util;
 import net.minecraft.advancements.CriteriaTriggers;
+import net.minecraft.client.Minecraft;
 import net.minecraft.commands.arguments.EntityAnchorArgument;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.NonNullList;
 import net.minecraft.core.SectionPos;
+import net.minecraft.core.particles.ItemParticleOption;
+import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.NbtOps;
 import net.minecraft.network.chat.ChatType;
@@ -53,6 +57,7 @@
 import net.minecraft.network.protocol.game.ClientboundPlayerCombatEndPacket;
 import net.minecraft.network.protocol.game.ClientboundPlayerCombatEnterPacket;
 import net.minecraft.network.protocol.game.ClientboundPlayerCombatKillPacket;
+import net.minecraft.network.protocol.game.ClientboundPlayerInfoPacket;
 import net.minecraft.network.protocol.game.ClientboundPlayerLookAtPacket;
 import net.minecraft.network.protocol.game.ClientboundRemoveMobEffectPacket;
 import net.minecraft.network.protocol.game.ClientboundResourcePackPacket;
@@ -109,7 +114,10 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
 import net.minecraft.world.item.ServerItemCooldowns;
+import net.minecraft.world.item.UseAnim;
 import net.minecraft.world.item.WrittenBookItem;
+import net.minecraft.world.item.alchemy.PotionUtils;
+import net.minecraft.world.item.alchemy.Potions;
 import net.minecraft.world.item.crafting.Recipe;
 import net.minecraft.world.item.trading.MerchantOffers;
 import net.minecraft.world.level.ChunkPos;
@@ -132,8 +140,12 @@
 import net.minecraft.world.scores.Score;
 import net.minecraft.world.scores.Team;
 import net.minecraft.world.scores.criteria.ObjectiveCriteria;
+import net.minecraftforge.common.util.ITeleporter;
+import net.optifine.reflect.Reflector;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.vivecraft.api.NetworkHelper;
+import org.vivecraft.api.ServerVivePlayer;
 
 public class ServerPlayer extends Player
 {
@@ -225,6 +237,9 @@
     private int containerCounter;
     public int latency;
     public boolean wonGame;
+    private String language = "en_us";
+    private boolean hasTabListName = false;
+    private Component tabListDisplayName = null;
 
     public ServerPlayer(MinecraftServer p_143384_, ServerLevel p_143385_, GameProfile p_143386_)
     {
@@ -370,9 +385,9 @@
             pCompound.putInt("SpawnZ", this.respawnPosition.getZ());
             pCompound.putBoolean("SpawnForced", this.respawnForced);
             pCompound.putFloat("SpawnAngle", this.respawnAngle);
-            ResourceLocation.CODEC.encodeStart(NbtOps.INSTANCE, this.respawnDimension.location()).resultOrPartial(LOGGER::error).ifPresent((p_9134_) ->
+            ResourceLocation.CODEC.encodeStart(NbtOps.INSTANCE, this.respawnDimension.location()).resultOrPartial(LOGGER::error).ifPresent((p_241148_1_) ->
             {
-                pCompound.put("SpawnDimension", p_9134_);
+                pCompound.put("SpawnDimension", p_241148_1_);
             });
         }
     }
@@ -405,13 +420,38 @@
 
     private void initMenu(AbstractContainerMenu p_143400_)
     {
-        p_143400_.addSlotListener(this.containerListener);
-        p_143400_.setSynchronizer(this.containerSynchronizer);
+    	p_143400_.addSlotListener(this.containerListener);
+    	p_143400_.setSynchronizer(this.containerSynchronizer);
     }
 
     public void initInventoryMenu()
     {
         this.initMenu(this.inventoryMenu);
+
+        ServerVivePlayer serverviveplayer = NetworkHelper.vivePlayers.get(this.getUUID());
+
+        //vivecraft
+        if (!Minecraft.getInstance().vrSettings.disableFun && serverviveplayer != null && serverviveplayer.isVR() && this.random.nextInt(40) == 3)
+        {
+            ItemStack itemstack;
+
+            if (this.random.nextInt(2) == 1)
+            {
+                itemstack = (new ItemStack(Items.PUMPKIN_PIE)).setHoverName(new TextComponent("EAT ME"));
+            }
+            else
+            {
+                itemstack = PotionUtils.setPotion(new ItemStack(Items.POTION), Potions.WATER).setHoverName(new TextComponent("DRINK ME"));
+            }
+
+            itemstack.getTag().putInt("HideFlags", 32);
+
+            if (this.getInventory().add(itemstack))
+            {
+                this.inventoryMenu.broadcastChanges();
+            }
+        }
+        //
     }
 
     public void onEnterCombat()
@@ -491,6 +531,7 @@
             if (!this.isSpectator() || !this.touchingUnloadedChunk())
             {
                 super.tick();
+                NetworkHelper.overridePose(this);
             }
 
             for (int i = 0; i < this.getInventory().getContainerSize(); ++i)
@@ -563,9 +604,9 @@
                 CriteriaTriggers.LOCATION.trigger(this);
             }
         }
-        catch (Throwable throwable)
+        catch (Throwable throwable1)
         {
-            CrashReport crashreport = CrashReport.forThrowable(throwable, "Ticking player");
+            CrashReport crashreport = CrashReport.forThrowable(throwable1, "Ticking player");
             CrashReportCategory crashreportcategory = crashreport.addCategory("Player being ticked");
             this.fillCrashReportCategory(crashreportcategory);
             throw new ReportedException(crashreport);
@@ -777,76 +818,110 @@
         }
     }
 
+    //Forge
     @Nullable
     public Entity changeDimension(ServerLevel pServer)
     {
-        this.isChangingDimension = true;
-        ServerLevel serverlevel = this.getLevel();
-        ResourceKey<Level> resourcekey = serverlevel.dimension();
-
-        if (resourcekey == Level.END && pServer.dimension() == Level.OVERWORLD)
-        {
-            this.unRide();
-            this.getLevel().removePlayerImmediately(this, Entity.RemovalReason.CHANGED_DIMENSION);
-
-            if (!this.wonGame)
-            {
-                this.wonGame = true;
-                this.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.WIN_GAME, this.seenCredits ? 0.0F : 1.0F));
-                this.seenCredits = true;
-            }
-
-            return this;
-        }
-        else
-        {
-            LevelData leveldata = pServer.getLevelData();
-            this.connection.send(new ClientboundRespawnPacket(pServer.dimensionType(), pServer.dimension(), BiomeManager.obfuscateSeed(pServer.getSeed()), this.gameMode.getGameModeForPlayer(), this.gameMode.getPreviousGameModeForPlayer(), pServer.isDebug(), pServer.isFlat(), true));
-            this.connection.send(new ClientboundChangeDifficultyPacket(leveldata.getDifficulty(), leveldata.isDifficultyLocked()));
-            PlayerList playerlist = this.server.getPlayerList();
-            playerlist.sendPlayerPermissionLevel(this);
-            serverlevel.removePlayerImmediately(this, Entity.RemovalReason.CHANGED_DIMENSION);
-            this.unsetRemoved();
-            PortalInfo portalinfo = this.findDimensionEntryPoint(pServer);
-
-            if (portalinfo != null)
-            {
-                serverlevel.getProfiler().push("moving");
-
-                if (resourcekey == Level.OVERWORLD && pServer.dimension() == Level.NETHER)
-                {
-                    this.enteredNetherPosition = this.position();
-                }
-                else if (pServer.dimension() == Level.END)
-                {
-                    this.createEndPlatform(pServer, new BlockPos(portalinfo.pos));
-                }
-
-                serverlevel.getProfiler().pop();
-                serverlevel.getProfiler().push("placing");
-                this.setLevel(pServer);
-                pServer.addDuringPortalTeleport(this);
-                this.setRot(portalinfo.yRot, portalinfo.xRot);
-                this.moveTo(portalinfo.pos.x, portalinfo.pos.y, portalinfo.pos.z);
-                serverlevel.getProfiler().pop();
-                this.triggerDimensionChangeTriggers(serverlevel);
-                this.connection.send(new ClientboundPlayerAbilitiesPacket(this.getAbilities()));
-                playerlist.sendLevelInfo(this, pServer);
-                playerlist.sendAllPlayerInfo(this);
-
-                for (MobEffectInstance mobeffectinstance : this.getActiveEffects())
-                {
-                    this.connection.send(new ClientboundUpdateMobEffectPacket(this.getId(), mobeffectinstance));
-                }
-
-                this.connection.send(new ClientboundLevelEventPacket(1032, BlockPos.ZERO, 0, false));
-                this.lastSentExp = -1;
-                this.lastSentHealth = -1.0F;
-                this.lastSentFood = -1;
-            }
-
-            return this;
-        }
+         return this.changeDimension(pServer, Reflector.ForgeHooks.exists() ? (ITeleporter)pServer.getPortalForcer() : null);
+     }
+    
+    @Nullable
+    public Entity changeDimension(ServerLevel server, ITeleporter teleporter)
+    {
+    	if (Reflector.ForgeHooks.exists() && !Reflector.callBoolean(Reflector.ForgeHooks_onTravelToDimension, this, server.dimension()))
+    	{
+    		return null;
+    	}
+    	else
+    	{
+    		this.isChangingDimension = true;
+    		ServerLevel serverlevel = this.getLevel();
+    		ResourceKey<Level> resourcekey = serverlevel.dimension();
+
+    		if (resourcekey == Level.END && server.dimension() == Level.OVERWORLD)
+    		{
+    			this.unRide();
+
+    			if (Reflector.ForgeHooks.exists())
+    			{
+    				Reflector.callVoid(this.getLevel(), Reflector.ServerWorld_removePlayer, this, true);
+    			}
+    			else
+    			{
+    				this.getLevel().removePlayerImmediately(this, RemovalReason.CHANGED_DIMENSION);
+    			}
+
+    			if (!this.wonGame)
+    			{
+    				this.wonGame = true;
+    				this.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.WIN_GAME, this.seenCredits ? 0.0F : 1.0F));
+    				this.seenCredits = true;
+    			}
+
+    			return this;
+    		}
+    		else
+    		{
+    			LevelData leveldata = server.getLevelData();
+    			this.connection.send(new ClientboundRespawnPacket(server.dimensionType(), server.dimension(), BiomeManager.obfuscateSeed(server.getSeed()), this.gameMode.getGameModeForPlayer(), this.gameMode.getPreviousGameModeForPlayer(), server.isDebug(), server.isFlat(), true));
+    			this.connection.send(new ClientboundChangeDifficultyPacket(leveldata.getDifficulty(), leveldata.isDifficultyLocked()));
+    			PlayerList playerlist = this.server.getPlayerList();
+    			playerlist.sendPlayerPermissionLevel(this);
+
+    			if (Reflector.ForgeHooks.exists())
+    			{
+    				Reflector.callVoid(serverlevel, Reflector.ServerWorld_removePlayer, this, true);
+    				Reflector.callVoid(this, Reflector.Entity_revive);
+    			}
+    			else
+    			{
+    				serverlevel.removePlayerImmediately(this, RemovalReason.CHANGED_DIMENSION);
+    				this.unsetRemoved();
+    			}
+
+    			PortalInfo portalinfo = this.findDimensionEntryPoint(server);
+
+    			if (portalinfo != null)
+    			{
+    				serverlevel.getProfiler().push("moving");
+
+    				if (resourcekey == Level.OVERWORLD && server.dimension() == Level.NETHER)
+    				{
+    					this.enteredNetherPosition = this.position();
+    				}
+    				else if (server.dimension() == Level.END)
+    				{
+    					this.createEndPlatform(server, new BlockPos(portalinfo.pos));
+    				}
+
+    				serverlevel.getProfiler().pop();
+    				serverlevel.getProfiler().push("placing");
+    				this.setLevel(server);
+    				server.addDuringPortalTeleport(this);
+    				this.setRot(portalinfo.yRot, portalinfo.xRot);
+    				this.moveTo(portalinfo.pos.x, portalinfo.pos.y, portalinfo.pos.z);
+    				serverlevel.getProfiler().pop();
+    				this.triggerDimensionChangeTriggers(serverlevel);
+    				this.gameMode.setLevel(server);
+    				this.connection.send(new ClientboundPlayerAbilitiesPacket(this.getAbilities()));
+    				playerlist.sendLevelInfo(this, server);
+    				playerlist.sendAllPlayerInfo(this);
+
+    				for (MobEffectInstance mobeffectinstance : this.getActiveEffects())
+    				{
+    					this.connection.send(new ClientboundUpdateMobEffectPacket(this.getId(), mobeffectinstance));
+    				}
+
+    				this.connection.send(new ClientboundLevelEventPacket(1032, BlockPos.ZERO, 0, false));
+    				this.lastSentExp = -1;
+    				this.lastSentHealth = -1.0F;
+    				this.lastSentFood = -1;
+    				Reflector.call(Reflector.BasicEventHooks_firePlayerChangedDimensionEvent, this, resourcekey, server.dimension());
+    			}
+
+    			return this;
+    		}
+    	}
     }
 
     private void createEndPlatform(ServerLevel p_9007_, BlockPos p_9008_)
@@ -938,6 +1013,18 @@
 
     public Either<Player.BedSleepingProblem, Unit> startSleepInBed(BlockPos pAt)
     {
+        Optional<BlockPos> optional = Optional.of(pAt);
+
+        if (Reflector.ForgeEventFactory_onPlayerSleepInBed.exists())
+        {
+            Player.BedSleepingProblem player$bedsleepingproblem = (Player.BedSleepingProblem)Reflector.call(Reflector.ForgeEventFactory_onPlayerSleepInBed, this, optional);
+
+            if (player$bedsleepingproblem != null)
+            {
+                return Either.left(player$bedsleepingproblem);
+            }
+        }
+
         Direction direction = this.level.getBlockState(pAt).getValue(HorizontalDirectionalBlock.FACING);
 
         if (!this.isSleeping() && this.isAlive())
@@ -957,8 +1044,14 @@
             else
             {
                 this.setRespawnPosition(this.level.dimension(), pAt, this.getYRot(), false, true);
-
-                if (this.level.isDay())
+                
+                boolean flag = this.level.isDay();
+                if (Reflector.ForgeEventFactory_fireSleepingTimeCheck.exists())
+                {
+                	flag = !Reflector.callBoolean(Reflector.ForgeEventFactory_fireSleepingTimeCheck, this, optional);
+                }
+                
+                if (flag)
                 {
                     return Either.left(Player.BedSleepingProblem.NOT_POSSIBLE_NOW);
                 }
@@ -969,9 +1062,9 @@
                         double d0 = 8.0D;
                         double d1 = 5.0D;
                         Vec3 vec3 = Vec3.atBottomCenterOf(pAt);
-                        List<Monster> list = this.level.getEntitiesOfClass(Monster.class, new AABB(vec3.x() - 8.0D, vec3.y() - 5.0D, vec3.z() - 8.0D, vec3.x() + 8.0D, vec3.y() + 5.0D, vec3.z() + 8.0D), (p_9062_) ->
+                        List<Monster> list = this.level.getEntitiesOfClass(Monster.class, new AABB(vec3.x() - 8.0D, vec3.y() - 5.0D, vec3.z() - 8.0D, vec3.x() + 8.0D, vec3.y() + 5.0D, vec3.z() + 8.0D), (p_241146_1_) ->
                         {
-                            return p_9062_.isPreventingPlayerRest(this);
+                            return p_241146_1_.isPreventingPlayerRest(this);
                         });
 
                         if (!list.isEmpty())
@@ -980,7 +1073,7 @@
                         }
                     }
 
-                    Either<Player.BedSleepingProblem, Unit> either = super.startSleepInBed(pAt).ifRight((p_9029_) ->
+                    Either<Player.BedSleepingProblem, Unit> either = super.startSleepInBed(pAt).ifRight((p_241144_1_) ->
                     {
                         this.awardStat(Stats.SLEEP_IN_BED);
                         CriteriaTriggers.SLEPT_IN_BED.trigger(this);
@@ -1010,7 +1103,14 @@
 
     private boolean bedInRange(BlockPos p_9117_, Direction p_9118_)
     {
-        return this.isReachableBedBlock(p_9117_) || this.isReachableBedBlock(p_9117_.relative(p_9118_.getOpposite()));
+        if (p_9118_ == null)
+        {
+            return false;
+        }
+        else
+        {
+            return this.isReachableBedBlock(p_9117_) || this.isReachableBedBlock(p_9117_.relative(p_9118_.getOpposite()));
+        }
     }
 
     private boolean isReachableBedBlock(BlockPos p_9223_)
@@ -1151,6 +1251,7 @@
                 this.connection.send(new ClientboundOpenScreenPacket(abstractcontainermenu.containerId, abstractcontainermenu.getType(), p_9033_.getDisplayName()));
                 this.initMenu(abstractcontainermenu);
                 this.containerMenu = abstractcontainermenu;
+                Reflector.postForgeBusEvent(Reflector.newInstance(Reflector.PlayerContainerEvent_Open_Constructor, this, this.containerMenu));
                 return OptionalInt.of(this.containerCounter);
             }
         }
@@ -1172,6 +1273,7 @@
         this.connection.send(new ClientboundHorseScreenOpenPacket(this.containerCounter, pInventory.getContainerSize(), pHorse.getId()));
         this.containerMenu = new HorseInventoryMenu(this.containerCounter, this.getInventory(), pInventory, pHorse);
         this.initMenu(this.containerMenu);
+        Reflector.postForgeBusEvent(Reflector.newInstance(Reflector.PlayerContainerEvent_Open_Constructor, this, this.containerMenu));
     }
 
     public void openItemGui(ItemStack pStack, InteractionHand pHand)
@@ -1203,6 +1305,7 @@
     {
         this.containerMenu.removed(this);
         this.inventoryMenu.transferState(this.containerMenu);
+        Reflector.postForgeBusEvent(Reflector.newInstance(Reflector.PlayerContainerEvent_Close_Constructor, this, this.containerMenu));
         this.containerMenu = this.inventoryMenu;
     }
 
@@ -1228,9 +1331,9 @@
     public void awardStat(Stat<?> pStat, int pAmount)
     {
         this.stats.increment(this, pStat, pAmount);
-        this.getScoreboard().forAllObjectives(pStat, this.getScoreboardName(), (p_8996_) ->
+        this.getScoreboard().forAllObjectives(pStat, this.getScoreboardName(), (p_195396_1_) ->
         {
-            p_8996_.add(pAmount);
+            p_195396_1_.add(pAmount);
         });
     }
 
@@ -1352,6 +1455,14 @@
         this.enteredNetherPosition = pThat.enteredNetherPosition;
         this.setShoulderEntityLeft(pThat.getShoulderEntityLeft());
         this.setShoulderEntityRight(pThat.getShoulderEntityRight());
+        CompoundTag compoundtag = (CompoundTag)Reflector.call(pThat, Reflector.Entity_getPersistentData);
+
+        if (compoundtag != null && compoundtag.contains("PlayerPersisted"))
+        {
+            ((CompoundTag)Reflector.call(this, Reflector.Entity_getPersistentData)).put("PlayerPersisted", compoundtag.getCompound("PlayerPersisted"));
+        }
+
+        Reflector.callVoid(Reflector.ForgeEventFactory_onPlayerClone, this, pThat, !pKeepEverything);
     }
 
     protected void onEffectAdded(MobEffectInstance p_143393_, @Nullable Entity p_143394_)
@@ -1495,6 +1606,11 @@
         this.textFilteringEnabled = pPacket.isTextFilteringEnabled();
         this.getEntityData().set(DATA_PLAYER_MODE_CUSTOMISATION, (byte)pPacket.getModelCustomisation());
         this.getEntityData().set(DATA_PLAYER_MAIN_HAND, (byte)(pPacket.getMainHand() == HumanoidArm.LEFT ? 0 : 1));
+
+        if (Reflector.ForgeHooks.exists())
+        {
+            this.language = Reflector.callString(pPacket, Reflector.CClientSettingsPacket_getLanguage);
+        }
     }
 
     public boolean canChatInColor()
@@ -1606,7 +1722,13 @@
     @Nullable
     public Component getTabListDisplayName()
     {
-        return null;
+        if (!this.hasTabListName)
+        {
+            this.tabListDisplayName = (Component)Reflector.call(Reflector.ForgeEventFactory_getPlayerTabListDisplayName, this);
+            this.hasTabListName = true;
+        }
+
+        return this.tabListDisplayName;
     }
 
     public void swing(InteractionHand pHand)
@@ -1646,8 +1768,20 @@
             this.connection.send(new ClientboundRespawnPacket(pX.dimensionType(), pX.dimension(), BiomeManager.obfuscateSeed(pX.getSeed()), this.gameMode.getGameModeForPlayer(), this.gameMode.getPreviousGameModeForPlayer(), pX.isDebug(), pX.isFlat(), true));
             this.connection.send(new ClientboundChangeDifficultyPacket(leveldata.getDifficulty(), leveldata.isDifficultyLocked()));
             this.server.getPlayerList().sendPlayerPermissionLevel(this);
-            serverlevel.removePlayerImmediately(this, Entity.RemovalReason.CHANGED_DIMENSION);
-            this.unsetRemoved();
+
+            //Forge
+            if (Reflector.ForgeHooks.exists())
+            {
+            	Reflector.callVoid(serverlevel, Reflector.ServerWorld_removePlayer, this, true);
+            	Reflector.callVoid(this, Reflector.Entity_revive);
+            }
+            else
+            {
+                serverlevel.removePlayerImmediately(this, Entity.RemovalReason.CHANGED_DIMENSION);
+                this.unsetRemoved();
+            }
+            //
+            
             this.moveTo(p_9001_, pY, p_9003_, pZ, p_9005_);
             this.setLevel(pX);
             pX.addDuringCommandTeleport(this);
@@ -1655,6 +1789,80 @@
             this.connection.teleport(p_9001_, pY, p_9003_, pZ, p_9005_);
             this.server.getPlayerList().sendLevelInfo(this, pX);
             this.server.getPlayerList().sendAllPlayerInfo(this);
+            Reflector.call(Reflector.BasicEventHooks_firePlayerChangedDimensionEvent, this, serverlevel.dimension(), pX.dimension());
+        }
+    }
+
+    public void sweepAttack()
+    {
+        ServerVivePlayer serverviveplayer = NetworkHelper.vivePlayers.get(this.getUUID());
+
+        if (serverviveplayer != null && serverviveplayer.isVR())
+        {
+            Vec3 vec3 = serverviveplayer.getControllerDir(0);
+            float f = (float)Math.toDegrees(Math.atan2(vec3.x, -vec3.z));
+            double d0 = (double)(-Mth.sin(f * ((float)Math.PI / 180F)));
+            double d1 = (double)Mth.cos(f * ((float)Math.PI / 180F));
+            Vec3 vec31 = serverviveplayer.getControllerPos(0, this);
+
+            if (this.level instanceof ServerLevel)
+            {
+                ((ServerLevel)this.level).sendParticles(ParticleTypes.SWEEP_ATTACK, vec31.x + d0, vec31.y, vec31.z + d1, 0, d0, 0.0D, d1, 0.0D);
+            }
+        }
+        else
+        {
+            super.sweepAttack();
+        }
+    }
+
+    protected void triggerItemUseEffects(ItemStack pStack, int pCount)
+    {
+        if (!pStack.isEmpty() && this.isUsingItem())
+        {
+            if (pStack.getUseAnimation() == UseAnim.DRINK)
+            {
+                this.playSound(this.getDrinkingSound(pStack), 0.5F, this.level.random.nextFloat() * 0.1F + 0.9F);
+            }
+
+            if (pStack.getUseAnimation() == UseAnim.EAT)
+            {
+                this.addItemParticles(pStack, pCount);
+                this.playSound(this.getEatingSound(pStack), 0.5F + 0.5F * (float)this.random.nextInt(2), (this.random.nextFloat() - this.random.nextFloat()) * 0.2F + 1.0F);
+            }
+        }
+    }
+
+    private void addItemParticles(ItemStack stack, int count)
+    {
+        ServerVivePlayer serverviveplayer = NetworkHelper.vivePlayers.get(this.getUUID());
+
+        for (int i = 0; i < count; ++i)
+        {
+            Vec3 vec3 = new Vec3(((double)this.random.nextFloat() - 0.5D) * 0.1D, Math.random() * 0.1D + 0.1D, 0.0D);
+            vec3 = vec3.xRot(-this.getXRot() * ((float)Math.PI / 180F));
+            vec3 = vec3.yRot(-this.getYRot() * ((float)Math.PI / 180F));
+            double d0 = (double)(-this.random.nextFloat()) * 0.6D - 0.3D;
+            Vec3 vec31 = new Vec3(((double)this.random.nextFloat() - 0.5D) * 0.3D, d0, 0.6D);
+            vec31 = vec31.xRot(-this.getXRot() * ((float)Math.PI / 180F));
+            vec31 = vec31.yRot(-this.getYRot() * ((float)Math.PI / 180F));
+            vec31 = vec31.add(this.getX(), this.getEyeY(), this.getZ());
+
+            if (serverviveplayer != null && serverviveplayer.isVR())
+            {
+                InteractionHand interactionhand = this.getUsedItemHand();
+
+                if (interactionhand == InteractionHand.MAIN_HAND)
+                {
+                    vec31 = serverviveplayer.getControllerPos(0, this);
+                }
+                else
+                {
+                    vec31 = serverviveplayer.getControllerPos(1, this);
+                }
+            }
+
+            this.level.addParticle(new ItemParticleOption(ParticleTypes.ITEM, stack), vec31.x, vec31.y, vec31.z, vec3.x, vec3.y + 0.05D, vec3.z);
         }
     }
 
@@ -1681,26 +1889,29 @@
 
     public void setRespawnPosition(ResourceKey<Level> p_9159_, @Nullable BlockPos p_9160_, float p_9161_, boolean p_9162_, boolean p_9163_)
     {
-        if (p_9160_ != null)
+        if (!Reflector.callBoolean(Reflector.ForgeEventFactory_onPlayerSpawnSet, this, p_9160_ == null ? Level.OVERWORLD : p_9159_, p_9160_, p_9161_))
         {
-            boolean flag = p_9160_.equals(this.respawnPosition) && p_9159_.equals(this.respawnDimension);
+            if (p_9160_ != null)
+            {
+                boolean flag = p_9160_.equals(this.respawnPosition) && p_9159_.equals(this.respawnDimension);
 
-            if (p_9163_ && !flag)
+                if (p_9163_ && !flag)
+                {
+                    this.sendMessage(new TranslatableComponent("block.minecraft.set_spawn"), Util.NIL_UUID);
+                }
+
+                this.respawnPosition = p_9160_;
+                this.respawnDimension = p_9159_;
+                this.respawnAngle = p_9161_;
+                this.respawnForced = p_9162_;
+            }
+            else
             {
-                this.sendMessage(new TranslatableComponent("block.minecraft.set_spawn"), Util.NIL_UUID);
+                this.respawnPosition = null;
+                this.respawnDimension = Level.OVERWORLD;
+                this.respawnAngle = 0.0F;
+                this.respawnForced = false;
             }
-
-            this.respawnPosition = p_9160_;
-            this.respawnDimension = p_9159_;
-            this.respawnAngle = p_9161_;
-            this.respawnForced = p_9162_;
-        }
-        else
-        {
-            this.respawnPosition = null;
-            this.respawnDimension = Level.OVERWORLD;
-            this.respawnAngle = 0.0F;
-            this.respawnForced = false;
         }
     }
 
@@ -1748,7 +1959,28 @@
         }
         else
         {
-            this.level.addFreshEntity(itementity);
+            ServerVivePlayer serverviveplayer = NetworkHelper.vivePlayers.get(this.getUUID());
+
+            if (serverviveplayer != null && serverviveplayer.isVR() && !pDropAround)
+            {
+                Vec3 vec3 = serverviveplayer.getControllerPos(0, this);
+                Vec3 vec31 = serverviveplayer.getControllerDir(0);
+                float f = 0.3F;
+                itementity.setDeltaMovement(vec31.x * (double)f, vec31.y * (double)f, vec31.z * (double)f);
+                itementity.setPos(vec3.x() + itementity.getDeltaMovement().x(), vec3.y() + itementity.getDeltaMovement().y(), vec3.z() + itementity.getDeltaMovement().z());
+            }
+
+            Collection<ItemEntity> collection = (Collection)Reflector.call(this, Reflector.Entity_captureDrops);
+
+            if (collection != null)
+            {
+                collection.add(itementity);
+            }
+            else
+            {
+                this.level.addFreshEntity(itementity);
+            }
+
             ItemStack itemstack = itementity.getItem();
 
             if (pTraceItem)
@@ -1850,4 +2082,21 @@
         });
         return this.drop(itemstack, false, true) != null;
     }
+
+    public String getLanguage()
+    {
+        return this.language;
+    }
+
+    public void refreshTabListName()
+    {
+        Component component = this.tabListDisplayName;
+        this.tabListDisplayName = (Component)Reflector.call(Reflector.ForgeEventFactory_getPlayerTabListDisplayName, this);
+
+        if (!Objects.equals(component, this.tabListDisplayName))
+        {
+            this.getServer().getPlayerList().broadcastAll(new ClientboundPlayerInfoPacket(ClientboundPlayerInfoPacket.Action.UPDATE_DISPLAY_NAME, this));
+        }
+    }
+
 }
