--- a/net/minecraft/client/Option.java
+++ b/net/minecraft/client/Option.java
@@ -435,6 +435,9 @@
             p_193661_0_.graphicsMode = p_193661_2_;
             p_193661_0_.updateRenderClouds();
             minecraft.levelRenderer.allChanged();
+            //Vivecraft
+            minecraft.vrRenderer.reinitFrameBuffers("gfx setting change");
+            //
         }
     }).setTooltip((p_193632_0_) ->
     {
