--- a/net/minecraft/client/Option.java
+++ b/net/minecraft/client/Option.java
@@ -377,6 +377,9 @@
             p_168176_0_.graphicsMode = p_168176_2_;
             p_168176_0_.updateRenderClouds();
             minecraft.levelRenderer.allChanged();
+            //Vivecraft
+            minecraft.vrRenderer.reinitFrameBuffers("gfx setting change");
+            //
         }
     }).setTooltip((p_168150_0_) ->
     {
