--- a/net/minecraft/client/Camera.java
+++ b/net/minecraft/client/Camera.java
@@ -21,17 +21,17 @@
 
 public class Camera
 {
-    private boolean initialized;
-    private BlockGetter level;
-    private Entity entity;
+    protected boolean initialized;
+    protected BlockGetter level;
+    protected Entity entity;
     private Vec3 position = Vec3.ZERO;
     private final BlockPos.MutableBlockPos blockPosition = new BlockPos.MutableBlockPos();
-    private final Vector3f forwards = new Vector3f(0.0F, 0.0F, 1.0F);
-    private final Vector3f up = new Vector3f(0.0F, 1.0F, 0.0F);
-    private final Vector3f left = new Vector3f(1.0F, 0.0F, 0.0F);
-    private float xRot;
-    private float yRot;
-    private final Quaternion rotation = new Quaternion(0.0F, 0.0F, 0.0F, 1.0F);
+    protected final Vector3f forwards = new Vector3f(0.0F, 0.0F, 1.0F);
+    protected final Vector3f up = new Vector3f(0.0F, 1.0F, 0.0F);
+    protected final Vector3f left = new Vector3f(1.0F, 0.0F, 0.0F);
+    protected float xRot;
+    protected float yRot;
+    protected final Quaternion rotation = new Quaternion(0.0F, 0.0F, 0.0F, 1.0F);
     private boolean detached;
     private float eyeHeight;
     private float eyeHeightOld;
@@ -238,10 +238,10 @@
         return !this.initialized ? Blocks.AIR.defaultBlockState() : this.level.getBlockState(this.blockPosition);
     }
 
-    public void setAnglesInternal(float yaw, float pitch)
+    public void setAnglesInternal(float p_setAnglesInternal_1_, float p_setAnglesInternal_2_)
     {
-        this.yRot = yaw;
-        this.xRot = pitch;
+        this.yRot = p_setAnglesInternal_1_;
+        this.xRot = p_setAnglesInternal_2_;
     }
 
     public BlockState getBlockAtCamera()
