--- a/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
+++ b/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
@@ -13,6 +13,9 @@
 import net.minecraft.world.level.Level;
 import net.minecraft.world.phys.HitResult;
 import net.minecraft.world.phys.Vec3;
+import net.optifine.reflect.Reflector;
+import org.vivecraft.api.NetworkHelper;
+import org.vivecraft.api.ServerVivePlayer;
 
 public abstract class AbstractHurtingProjectile extends Projectile
 {
@@ -78,8 +81,14 @@
             }
 
             HitResult hitresult = ProjectileUtil.getHitResult(this, this::canHitEntity);
+            boolean flag = false;
 
-            if (hitresult.getType() != HitResult.Type.MISS)
+            if (Reflector.ForgeEventFactory_onProjectileImpactD.exists())
+            {
+                flag = Reflector.callBoolean(Reflector.ForgeEventFactory_onProjectileImpactD, this, hitresult);
+            }
+
+            if (hitresult.getType() != HitResult.Type.MISS && !flag)
             {
                 this.onHit(hitresult);
             }
@@ -180,6 +189,13 @@
             if (entity != null)
             {
                 Vec3 vec3 = entity.getLookAngle();
+                ServerVivePlayer serverviveplayer = NetworkHelper.vivePlayers.get(pSource.getEntity().getUUID());
+
+                if (serverviveplayer != null && serverviveplayer.isVR())
+                {
+                    vec3 = serverviveplayer.getHMDDir();
+                }
+
                 this.setDeltaMovement(vec3);
                 this.xPower = vec3.x * 0.1D;
                 this.yPower = vec3.y * 0.1D;
