--- a/net/minecraft/client/gui/screens/inventory/CreativeModeInventoryScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/CreativeModeInventoryScreen.java
@@ -47,6 +47,7 @@
 import net.minecraft.world.item.TooltipFlag;
 import net.minecraft.world.item.enchantment.Enchantment;
 import net.minecraft.world.item.enchantment.EnchantmentHelper;
+import org.vivecraft.asm.ASMDelegator;
 
 public class CreativeModeInventoryScreen extends EffectRenderingInventoryScreen<CreativeModeInventoryScreen.ItemPickerMenu>
 {
@@ -464,6 +465,7 @@
             (this.menu).items.addAll(searchtree.search(s.toLowerCase(Locale.ROOT)));
         }
 
+        ASMDelegator.addCreativeSearch(this.searchBox.getValue(), (this.menu).items);
         this.scrollOffs = 0.0F;
         this.menu.scrollTo(0.0F);
     }
@@ -602,6 +604,7 @@
         else if (pTab != CreativeModeTab.TAB_SEARCH)
         {
             pTab.fillItemList((this.menu).items);
+            ASMDelegator.addCreativeItems(pTab, (this.menu).items);
         }
 
         if (pTab == CreativeModeTab.TAB_INVENTORY)
