--- a/net/minecraft/client/gui/screens/inventory/SignEditScreen.java
+++ b/net/minecraft/client/gui/screens/inventory/SignEditScreen.java
@@ -12,6 +12,9 @@
 import com.mojang.blaze3d.vertex.VertexFormat;
 import com.mojang.math.Matrix4f;
 import java.util.stream.IntStream;
+
+import org.vivecraft.gameplay.screenhandlers.KeyboardHandler;
+
 import net.minecraft.client.gui.components.Button;
 import net.minecraft.client.gui.font.TextFieldHelper;
 import net.minecraft.client.gui.screens.Screen;
@@ -63,6 +66,9 @@
             this.onDone();
         }));
         this.sign.setEditable(false);
+        //Vivecraft
+        KeyboardHandler.setOverlayShowing(true);
+        //
         this.signField = new TextFieldHelper(() ->
         {
             return this.messages[this.line];
@@ -90,6 +96,7 @@
         }
 
         this.sign.setEditable(true);
+        KeyboardHandler.setOverlayShowing(false);
     }
 
     public void tick()
