--- a/net/minecraft/client/renderer/entity/EntityRenderer.java
+++ b/net/minecraft/client/renderer/entity/EntityRenderer.java
@@ -133,7 +133,7 @@
             int i = "deadmau5".equals(pDisplayName.getString()) ? -10 : 0;
             pMatrixStack.pushPose();
             pMatrixStack.translate(0.0D, (double)f, 0.0D);
-            pMatrixStack.mulPose(this.entityRenderDispatcher.cameraOrientation());
+            pMatrixStack.mulPose(this.entityRenderDispatcher.getCameraOrientationOffset(0.5F));
             pMatrixStack.scale(-0.025F, -0.025F, 0.025F);
             Matrix4f matrix4f = pMatrixStack.last().pose();
             float f1 = Minecraft.getInstance().options.getBackgroundOpacity(0.25F);
@@ -156,9 +156,9 @@
         return this.entityType == null ? null : Either.makeLeft(this.entityType);
     }
 
-    public void setType(Either<EntityType, BlockEntityType> type)
+    public void setType(Either<EntityType, BlockEntityType> p_setType_1_)
     {
-        this.entityType = type.getLeft().get();
+        this.entityType = p_setType_1_.getLeft().get();
     }
 
     public ResourceLocation getLocationTextureCustom()
@@ -166,8 +166,8 @@
         return this.locationTextureCustom;
     }
 
-    public void setLocationTextureCustom(ResourceLocation locationTextureCustom)
+    public void setLocationTextureCustom(ResourceLocation p_setLocationTextureCustom_1_)
     {
-        this.locationTextureCustom = locationTextureCustom;
+        this.locationTextureCustom = p_setLocationTextureCustom_1_;
     }
 }
