--- a/net/minecraft/client/gui/screens/ChatScreen.java
+++ b/net/minecraft/client/gui/screens/ChatScreen.java
@@ -126,8 +126,17 @@
             {
                 this.sendMessage(s);
             }
-
-            this.minecraft.setScreen((Screen)null);
+            //Vivecraft
+            if (this.minecraft.vrSettings.seated)
+            {
+                this.minecraft.setScreen((Screen)null);
+            }
+            else
+            {
+                this.input.setValue(this.initial);
+                this.historyPos = this.minecraft.gui.getChat().getRecentChat().size();
+            }
+            //
             return true;
         }
     }
