--- a/net/minecraft/client/gui/screens/SoundOptionsScreen.java
+++ b/net/minecraft/client/gui/screens/SoundOptionsScreen.java
@@ -6,8 +6,10 @@
 import net.minecraft.client.gui.components.Button;
 import net.minecraft.client.gui.components.VolumeSlider;
 import net.minecraft.network.chat.CommonComponents;
+import net.minecraft.network.chat.TextComponent;
 import net.minecraft.network.chat.TranslatableComponent;
 import net.minecraft.sounds.SoundSource;
+import org.vivecraft.settings.VRSettings;
 
 public class SoundOptionsScreen extends OptionsSubScreen
 {
@@ -30,7 +32,15 @@
                 ++i;
             }
         }
-
+        
+        this.addRenderableWidget(new Button(this.width / 2 - 155 + i % 2 * 160, this.height / 6 - 12 + 24 * (i >> 1), 150, 20, new TextComponent(this.minecraft.vrSettings.getButtonDisplayString(VRSettings.VrOptions.HRTF_SELECTION)), (p_213104_1_) ->
+        {
+        	this.clearWidgets();
+        	this.minecraft.vrSettings.setOptionValue(VRSettings.VrOptions.HRTF_SELECTION);
+        	this.minecraft.vrSettings.saveOptions();
+        	this.init();
+        }));
+        
         int j = this.width / 2 - 75;
         int k = this.height / 6 - 12;
         ++i;
